<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Preprocessing and clustering 3k PBMCs · Severo.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><script src="../../copy.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">Severo.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><span class="tocitem">Showcases</span><ul><li class="is-active"><a class="tocitem" href>Preprocessing and clustering 3k PBMCs</a><ul class="internal"><li><a class="tocitem" href="#Overview"><span>Overview</span></a></li><li><a class="tocitem" href="#Loading-data"><span>Loading data</span></a></li><li><a class="tocitem" href="#Preprocessing"><span>Preprocessing</span></a></li><li><a class="tocitem" href="#Dimensionality-reduction"><span>Dimensionality reduction</span></a></li><li><a class="tocitem" href="#Clustering-the-cells"><span>Clustering the cells</span></a></li><li><a class="tocitem" href="#Finding-differentially-expressed-features"><span>Finding differentially expressed features</span></a></li></ul></li><li><a class="tocitem" href="../performance/">Performance</a></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../public/">Public</a></li><li><a class="tocitem" href="../internals/">Internals</a></li></ul></li><li><a class="tocitem" href="../contributing/">Contributing</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Showcases</a></li><li class="is-active"><a href>Preprocessing and clustering 3k PBMCs</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Preprocessing and clustering 3k PBMCs</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/ExaScience/Severo.jl/blob/master/docs/src/pbmc.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Preprocessing-and-clustering-3k-PBMCs"><a class="docs-heading-anchor" href="#Preprocessing-and-clustering-3k-PBMCs">Preprocessing and clustering 3k PBMCs</a><a id="Preprocessing-and-clustering-3k-PBMCs-1"></a><a class="docs-heading-anchor-permalink" href="#Preprocessing-and-clustering-3k-PBMCs" title="Permalink"></a></h1><p>This showcase reproduces Seurat&#39;s <a href="http://satijalab.org/seurat/pbmc3k_tutorial.html">Guided Clustering Tutorial</a></p><h2 id="Overview"><a class="docs-heading-anchor" href="#Overview">Overview</a><a id="Overview-1"></a><a class="docs-heading-anchor-permalink" href="#Overview" title="Permalink"></a></h2><ul><li><a href="#Preprocessing-and-clustering-3k-PBMCs">Preprocessing and clustering 3k PBMCs</a></li><li class="no-marker"><ul><li><a href="#Overview">Overview</a></li><li><a href="#Loading-data">Loading data</a></li><li><a href="#Preprocessing">Preprocessing</a></li><li><a href="#Dimensionality-reduction">Dimensionality reduction</a></li><li><a href="#Clustering-the-cells">Clustering the cells</a></li><li><a href="#Finding-differentially-expressed-features">Finding differentially expressed features</a></li></ul></li></ul><h2 id="Loading-data"><a class="docs-heading-anchor" href="#Loading-data">Loading data</a><a id="Loading-data-1"></a><a class="docs-heading-anchor-permalink" href="#Loading-data" title="Permalink"></a></h2><p>The data consists of 3k PBMCs from a Healthy Donor and is freely available from <a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/1.1.0/pbmc3k">10x Genomics</a>.</p><p>You can either download the data manually</p><pre><code class="language-bash hljs">mkdir pbmc3k
wget http://cf.10xgenomics.com/samples/cell-exp/1.1.0/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz -O pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz
cd pbmc3k; tar -xzf pbmc3k_filtered_gene_bc_matrices.tar.gz --strip-components 2</code></pre><p>The <code>read_10X</code> function reads the data from the cellranger 10X pipeline and returns a labelled count matrix. Each entry indicated the number of molecules detected for each feature/gene (columns) and each cell (rows).</p><pre><code class="language-julia hljs">using Severo
X = read_10X(&quot;pbmc3k/&quot;)</code></pre><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>The meaning of the rows and columns is different from the representation used by other packages like <a href="https://satijalab.org/seurat/">Seurat</a></p></div></div><p>Alternatively, the <code>dataset</code> function can be used to load a dataset from a predefined collection. For example, the PBMC collection is known by <code>Cell.jl</code> and can be easily loaded as follows:</p><pre><code class="language-julia hljs">using Severo
X = dataset(&quot;PBMC&quot;, &quot;3k&quot;)

# The matrix can be indexed using names or indices. For instance,
# we can look at specific genes in the first thirty cells
#X[1:30, [&quot;CD3D&quot;, &quot;TCL1A&quot;, &quot;MS4A1&quot;]]</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">2700×32738 Named sparse matrix with 2286884 Int32 nonzero entries:
	[AGATTCCTGACGAG-1,          &quot;AL627309.1&quot;]  =  1
	[CCAAAGTGCTACGA-1,          &quot;AL627309.1&quot;]  =  1
	[CCGTACACGTTGGT-1,          &quot;AL627309.1&quot;]  =  1
	[CGACTGCTTCCTCG-1,          &quot;AL627309.1&quot;]  =  1
	[CTATGTTGTCTCGC-1,          &quot;AL627309.1&quot;]  =  1
	[CTGTGAGACGAACT-1,          &quot;AL627309.1&quot;]  =  1
	[GATACTCTATCGGT-1,          &quot;AL627309.1&quot;]  =  1
	[GCTAGATGAGCTCA-1,          &quot;AL627309.1&quot;]  =  1
	[GCTTAACTACTGGT-1,          &quot;AL627309.1&quot;]  =  1
	                                           ⋮
	[TCTCAAACCTAAGC-1,            &quot;SRSF10.1&quot;]  =  1
	[TGAGACACTCAAGC-1,            &quot;SRSF10.1&quot;]  =  1
	[TGATCGGATATGCG-1,            &quot;SRSF10.1&quot;]  =  1
	[TGTAGGTGCTATGG-1,            &quot;SRSF10.1&quot;]  =  1
	[TTATCCGAGAAAGT-1,            &quot;SRSF10.1&quot;]  =  1
	[TTCAGTACCGACTA-1,            &quot;SRSF10.1&quot;]  =  1
	[TTGAATGATCTCAT-1,            &quot;SRSF10.1&quot;]  =  1
	[TTGAGGACTACGCA-1,            &quot;SRSF10.1&quot;]  =  1
	[TTTATCCTGTTGTG-1,            &quot;SRSF10.1&quot;]  =  1
	[TTTCCAGAGGTGAG-1,            &quot;SRSF10.1&quot;]  =  1</code></pre><p>The count data is stored in a sparse matrix format, only storing non-zero elements of the matrix. Any values not shown are zero.</p><article class="docstring"><header><a class="docstring-binding" id="Severo.read_data" href="#Severo.read_data"><code>Severo.read_data</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">read_data(path::AbstractString; kw...)</code></pre><p>Tries to identify and read a count matrix in any of the supported formats</p><p><strong>Arguments</strong>:</p><ul><li><code>fname</code>: path</li><li><code>kw</code>: additional keyword arguments are passed on</li></ul><p><strong>Returns values</strong>:</p><p>Returns labeled sparse matrix containing the counts</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/862003e8b61d12b112e2e343b7f508c1a959ad97/src/input.jl#L720-L733">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Severo.read_10X" href="#Severo.read_10X"><code>Severo.read_10X</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">read_10X(dirname::AbstractString; unique_features=true)</code></pre><p>Read count matrix from 10X genomics</p><p><strong>Arguments</strong>:</p><ul><li><code>dirname</code>: path to directory containing matrix.mtx, genes.tsv (or features.tsv), and barcodes.tsv from 10X</li><li><code>unique_features</code>: should feature names be made unique (default: true)</li></ul><p><strong>Returns values</strong>:</p><p>Returns labeled sparse matrix containing the counts</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/862003e8b61d12b112e2e343b7f508c1a959ad97/src/input.jl#L620-L633">source</a></section></article><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>dataset</code>. Check Documenter&#39;s build log for details.</p></div></div><h2 id="Preprocessing"><a class="docs-heading-anchor" href="#Preprocessing">Preprocessing</a><a id="Preprocessing-1"></a><a class="docs-heading-anchor-permalink" href="#Preprocessing" title="Permalink"></a></h2><pre><code class="language-julia hljs">using Plots
using StatsPlots
plot_highest_expressed(X)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">┌ Warning: Package Severo does not have StatsPlots in its dependencies:
│ - If you have Severo checked out for development and have
│   added StatsPlots as a dependency but haven&#39;t updated your primary
│   environment&#39;s manifest file, try `Pkg.resolve()`.
│ - Otherwise you may need to report an issue with Severo
└ Loading StatsPlots into Severo from project dependency, future warnings for Severo are suppressed.
[ Info: Installing matplotlib via the Conda matplotlib package...
[ Info: Running `conda install -q -y matplotlib` in root environment
Collecting package metadata (current_repodata.json): ...working... done
Solving environment: ...working... done

## Package Plan ##

  environment location: /home/runner/.julia/conda/3

  added / updated specs:
    - matplotlib


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    alsa-lib-1.2.3             |       h516909a_0         560 KB  conda-forge
    brotli-1.0.9               |       h7f98852_6          18 KB  conda-forge
    brotli-bin-1.0.9           |       h7f98852_6          19 KB  conda-forge
    cycler-0.11.0              |     pyhd8ed1ab_0          10 KB  conda-forge
    dbus-1.13.6                |       h5008d03_3         604 KB  conda-forge
    expat-2.4.4                |       h9c3ff4c_0         183 KB  conda-forge
    font-ttf-dejavu-sans-mono-2.37|       hab24e00_0         388 KB  conda-forge
    font-ttf-inconsolata-3.000 |       h77eed37_0          94 KB  conda-forge
    font-ttf-source-code-pro-2.038|       h77eed37_0         684 KB  conda-forge
    font-ttf-ubuntu-0.83       |       hab24e00_0         1.9 MB  conda-forge
    fontconfig-2.13.94         |       ha180cfb_0         323 KB  conda-forge
    fonts-conda-ecosystem-1    |                0           4 KB  conda-forge
    fonts-conda-forge-1        |                0           4 KB  conda-forge
    fonttools-4.29.1           |   py39h3811e60_0         1.6 MB  conda-forge
    freetype-2.10.4            |       h0708190_1         890 KB  conda-forge
    gettext-0.19.8.1           |    h73d1719_1008         3.6 MB  conda-forge
    gst-plugins-base-1.18.5    |       hf529b03_3         2.6 MB  conda-forge
    gstreamer-1.18.5           |       h9f60fe5_3         2.0 MB  conda-forge
    icu-69.1                   |       h9c3ff4c_0        13.2 MB  conda-forge
    jbig-2.1                   |    h7f98852_2003          43 KB  conda-forge
    jpeg-9e                    |       h7f98852_0         269 KB  conda-forge
    kiwisolver-1.3.2           |   py39h1a9c180_1          80 KB  conda-forge
    krb5-1.19.2                |       hcc1bbae_3         1.4 MB  conda-forge
    lcms2-2.12                 |       hddcbb42_0         443 KB  conda-forge
    lerc-3.0                   |       h9c3ff4c_0         216 KB  conda-forge
    libbrotlicommon-1.0.9      |       h7f98852_6          65 KB  conda-forge
    libbrotlidec-1.0.9         |       h7f98852_6          33 KB  conda-forge
    libbrotlienc-1.0.9         |       h7f98852_6         286 KB  conda-forge
    libclang-13.0.0            |default_hc23dcda_0        11.1 MB  conda-forge
    libdeflate-1.8             |       h7f98852_0          67 KB  conda-forge
    libedit-3.1.20191231       |       he28a2e2_2         121 KB  conda-forge
    libevent-2.1.10            |       h9b69904_4         1.1 MB  conda-forge
    libglib-2.70.2             |       h174f98d_1         3.1 MB  conda-forge
    libiconv-1.16              |       h516909a_0         1.4 MB  conda-forge
    libllvm13-13.0.0           |       hf817b99_0        33.6 MB  conda-forge
    libogg-1.3.4               |       h7f98852_1         206 KB  conda-forge
    libopus-1.3.1              |       h7f98852_1         255 KB  conda-forge
    libpng-1.6.37              |       h21135ba_2         306 KB  conda-forge
    libpq-14.1                 |       hd57d9b9_1         3.0 MB  conda-forge
    libtiff-4.3.0              |       h6f004c6_2         614 KB  conda-forge
    libuuid-2.32.1             |    h7f98852_1000          28 KB  conda-forge
    libvorbis-1.3.7            |       h9c3ff4c_0         280 KB  conda-forge
    libwebp-base-1.2.2         |       h7f98852_1         824 KB  conda-forge
    libxcb-1.13                |    h7f98852_1004         391 KB  conda-forge
    libxkbcommon-1.0.3         |       he3ba5ed_0         581 KB  conda-forge
    libxml2-2.9.12             |       h885dcf4_1         760 KB  conda-forge
    lz4-c-1.9.3                |       h9c3ff4c_1         179 KB  conda-forge
    matplotlib-3.5.1           |   py39hf3d152e_0           6 KB  conda-forge
    matplotlib-base-3.5.1      |   py39h2fa2bec_0         7.4 MB  conda-forge
    munkres-1.1.4              |     pyh9f0ad1d_0          12 KB  conda-forge
    mysql-common-8.0.28        |       ha770c72_0         1.8 MB  conda-forge
    mysql-libs-8.0.28          |       hfa10184_0         1.9 MB  conda-forge
    nspr-4.32                  |       h9c3ff4c_1         233 KB  conda-forge
    nss-3.74                   |       hb5efdd6_0         2.1 MB  conda-forge
    olefile-0.46               |     pyh9f0ad1d_1          32 KB  conda-forge
    openjpeg-2.4.0             |       hb52868f_1         444 KB  conda-forge
    packaging-21.3             |     pyhd8ed1ab_0          36 KB  conda-forge
    pcre-8.45                  |       h9c3ff4c_0         253 KB  conda-forge
    pillow-8.4.0               |   py39ha612740_0         706 KB  conda-forge
    pthread-stubs-0.4          |    h36c2ea0_1001           5 KB  conda-forge
    pyparsing-3.0.7            |     pyhd8ed1ab_0          79 KB  conda-forge
    pyqt-5.12.3                |   py39hf3d152e_8          22 KB  conda-forge
    pyqt-impl-5.12.3           |   py39hde8b62d_8         5.9 MB  conda-forge
    pyqt5-sip-4.19.18          |   py39he80948d_8         311 KB  conda-forge
    pyqtchart-5.12             |   py39h0fcd23e_8         257 KB  conda-forge
    pyqtwebengine-5.12.1       |   py39h0fcd23e_8         174 KB  conda-forge
    python-dateutil-2.8.2      |     pyhd8ed1ab_0         240 KB  conda-forge
    qt-5.12.9                  |       ha98a1a1_5        98.8 MB  conda-forge
    tornado-6.1                |   py39h3811e60_2         646 KB  conda-forge
    unicodedata2-14.0.0        |   py39h3811e60_0         545 KB  conda-forge
    xorg-libxau-1.0.9          |       h7f98852_0          13 KB  conda-forge
    xorg-libxdmcp-1.1.3        |       h7f98852_0          19 KB  conda-forge
    zstd-1.5.2                 |       ha95c52a_0         458 KB  conda-forge
    ------------------------------------------------------------
                                           Total:       211.3 MB

The following NEW packages will be INSTALLED:

  alsa-lib           conda-forge/linux-64::alsa-lib-1.2.3-h516909a_0
  brotli             conda-forge/linux-64::brotli-1.0.9-h7f98852_6
  brotli-bin         conda-forge/linux-64::brotli-bin-1.0.9-h7f98852_6
  cycler             conda-forge/noarch::cycler-0.11.0-pyhd8ed1ab_0
  dbus               conda-forge/linux-64::dbus-1.13.6-h5008d03_3
  expat              conda-forge/linux-64::expat-2.4.4-h9c3ff4c_0
  font-ttf-dejavu-s~ conda-forge/noarch::font-ttf-dejavu-sans-mono-2.37-hab24e00_0
  font-ttf-inconsol~ conda-forge/noarch::font-ttf-inconsolata-3.000-h77eed37_0
  font-ttf-source-c~ conda-forge/noarch::font-ttf-source-code-pro-2.038-h77eed37_0
  font-ttf-ubuntu    conda-forge/noarch::font-ttf-ubuntu-0.83-hab24e00_0
  fontconfig         conda-forge/linux-64::fontconfig-2.13.94-ha180cfb_0
  fonts-conda-ecosy~ conda-forge/noarch::fonts-conda-ecosystem-1-0
  fonts-conda-forge  conda-forge/noarch::fonts-conda-forge-1-0
  fonttools          conda-forge/linux-64::fonttools-4.29.1-py39h3811e60_0
  freetype           conda-forge/linux-64::freetype-2.10.4-h0708190_1
  gettext            conda-forge/linux-64::gettext-0.19.8.1-h73d1719_1008
  gst-plugins-base   conda-forge/linux-64::gst-plugins-base-1.18.5-hf529b03_3
  gstreamer          conda-forge/linux-64::gstreamer-1.18.5-h9f60fe5_3
  icu                conda-forge/linux-64::icu-69.1-h9c3ff4c_0
  jbig               conda-forge/linux-64::jbig-2.1-h7f98852_2003
  jpeg               conda-forge/linux-64::jpeg-9e-h7f98852_0
  kiwisolver         conda-forge/linux-64::kiwisolver-1.3.2-py39h1a9c180_1
  krb5               conda-forge/linux-64::krb5-1.19.2-hcc1bbae_3
  lcms2              conda-forge/linux-64::lcms2-2.12-hddcbb42_0
  lerc               conda-forge/linux-64::lerc-3.0-h9c3ff4c_0
  libbrotlicommon    conda-forge/linux-64::libbrotlicommon-1.0.9-h7f98852_6
  libbrotlidec       conda-forge/linux-64::libbrotlidec-1.0.9-h7f98852_6
  libbrotlienc       conda-forge/linux-64::libbrotlienc-1.0.9-h7f98852_6
  libclang           conda-forge/linux-64::libclang-13.0.0-default_hc23dcda_0
  libdeflate         conda-forge/linux-64::libdeflate-1.8-h7f98852_0
  libedit            conda-forge/linux-64::libedit-3.1.20191231-he28a2e2_2
  libevent           conda-forge/linux-64::libevent-2.1.10-h9b69904_4
  libglib            conda-forge/linux-64::libglib-2.70.2-h174f98d_1
  libiconv           conda-forge/linux-64::libiconv-1.16-h516909a_0
  libllvm13          conda-forge/linux-64::libllvm13-13.0.0-hf817b99_0
  libogg             conda-forge/linux-64::libogg-1.3.4-h7f98852_1
  libopus            conda-forge/linux-64::libopus-1.3.1-h7f98852_1
  libpng             conda-forge/linux-64::libpng-1.6.37-h21135ba_2
  libpq              conda-forge/linux-64::libpq-14.1-hd57d9b9_1
  libtiff            conda-forge/linux-64::libtiff-4.3.0-h6f004c6_2
  libuuid            conda-forge/linux-64::libuuid-2.32.1-h7f98852_1000
  libvorbis          conda-forge/linux-64::libvorbis-1.3.7-h9c3ff4c_0
  libwebp-base       conda-forge/linux-64::libwebp-base-1.2.2-h7f98852_1
  libxcb             conda-forge/linux-64::libxcb-1.13-h7f98852_1004
  libxkbcommon       conda-forge/linux-64::libxkbcommon-1.0.3-he3ba5ed_0
  libxml2            conda-forge/linux-64::libxml2-2.9.12-h885dcf4_1
  lz4-c              conda-forge/linux-64::lz4-c-1.9.3-h9c3ff4c_1
  matplotlib         conda-forge/linux-64::matplotlib-3.5.1-py39hf3d152e_0
  matplotlib-base    conda-forge/linux-64::matplotlib-base-3.5.1-py39h2fa2bec_0
  munkres            conda-forge/noarch::munkres-1.1.4-pyh9f0ad1d_0
  mysql-common       conda-forge/linux-64::mysql-common-8.0.28-ha770c72_0
  mysql-libs         conda-forge/linux-64::mysql-libs-8.0.28-hfa10184_0
  nspr               conda-forge/linux-64::nspr-4.32-h9c3ff4c_1
  nss                conda-forge/linux-64::nss-3.74-hb5efdd6_0
  olefile            conda-forge/noarch::olefile-0.46-pyh9f0ad1d_1
  openjpeg           conda-forge/linux-64::openjpeg-2.4.0-hb52868f_1
  packaging          conda-forge/noarch::packaging-21.3-pyhd8ed1ab_0
  pcre               conda-forge/linux-64::pcre-8.45-h9c3ff4c_0
  pillow             conda-forge/linux-64::pillow-8.4.0-py39ha612740_0
  pthread-stubs      conda-forge/linux-64::pthread-stubs-0.4-h36c2ea0_1001
  pyparsing          conda-forge/noarch::pyparsing-3.0.7-pyhd8ed1ab_0
  pyqt               conda-forge/linux-64::pyqt-5.12.3-py39hf3d152e_8
  pyqt-impl          conda-forge/linux-64::pyqt-impl-5.12.3-py39hde8b62d_8
  pyqt5-sip          conda-forge/linux-64::pyqt5-sip-4.19.18-py39he80948d_8
  pyqtchart          conda-forge/linux-64::pyqtchart-5.12-py39h0fcd23e_8
  pyqtwebengine      conda-forge/linux-64::pyqtwebengine-5.12.1-py39h0fcd23e_8
  python-dateutil    conda-forge/noarch::python-dateutil-2.8.2-pyhd8ed1ab_0
  qt                 conda-forge/linux-64::qt-5.12.9-ha98a1a1_5
  tornado            conda-forge/linux-64::tornado-6.1-py39h3811e60_2
  unicodedata2       conda-forge/linux-64::unicodedata2-14.0.0-py39h3811e60_0
  xorg-libxau        conda-forge/linux-64::xorg-libxau-1.0.9-h7f98852_0
  xorg-libxdmcp      conda-forge/linux-64::xorg-libxdmcp-1.1.3-h7f98852_0
  zstd               conda-forge/linux-64::zstd-1.5.2-ha95c52a_0


Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done
┌ Warning: seriestype boxplot has been moved to StatsPlots.  To use: `Pkg.add(&quot;StatsPlots&quot;); using StatsPlots`
└ @ Plots ~/.julia/packages/Plots/Xv2qA/src/args.jl:1575
┌ Warning: seriestype boxplot has been moved to StatsPlots.  To use: `Pkg.add(&quot;StatsPlots&quot;); using StatsPlots`
└ @ Plots ~/.julia/packages/Plots/Xv2qA/src/args.jl:1575
┌ Warning: seriestype boxplot has been moved to StatsPlots.  To use: `Pkg.add(&quot;StatsPlots&quot;); using StatsPlots`
└ @ Plots ~/.julia/packages/Plots/Xv2qA/src/args.jl:1575
┌ Warning: seriestype boxplot has been moved to StatsPlots.  To use: `Pkg.add(&quot;StatsPlots&quot;); using StatsPlots`
└ @ Plots ~/.julia/packages/Plots/Xv2qA/src/args.jl:1575
┌ Warning: seriestype boxplot has been moved to StatsPlots.  To use: `Pkg.add(&quot;StatsPlots&quot;); using StatsPlots`
└ @ Plots ~/.julia/packages/Plots/Xv2qA/src/args.jl:1575
┌ Warning: seriestype boxplot has been moved to StatsPlots.  To use: `Pkg.add(&quot;StatsPlots&quot;); using StatsPlots`
└ @ Plots ~/.julia/packages/Plots/Xv2qA/src/args.jl:1575
┌ Warning: seriestype boxplot has been moved to StatsPlots.  To use: `Pkg.add(&quot;StatsPlots&quot;); using StatsPlots`
└ @ Plots ~/.julia/packages/Plots/Xv2qA/src/args.jl:1575
┌ Warning: seriestype boxplot has been moved to StatsPlots.  To use: `Pkg.add(&quot;StatsPlots&quot;); using StatsPlots`
└ @ Plots ~/.julia/packages/Plots/Xv2qA/src/args.jl:1575
┌ Warning: seriestype boxplot has been moved to StatsPlots.  To use: `Pkg.add(&quot;StatsPlots&quot;); using StatsPlots`
└ @ Plots ~/.julia/packages/Plots/Xv2qA/src/args.jl:1575
┌ Warning: seriestype boxplot has been moved to StatsPlots.  To use: `Pkg.add(&quot;StatsPlots&quot;); using StatsPlots`
└ @ Plots ~/.julia/packages/Plots/Xv2qA/src/args.jl:1575
┌ Warning: seriestype boxplot has been moved to StatsPlots.  To use: `Pkg.add(&quot;StatsPlots&quot;); using StatsPlots`
└ @ Plots ~/.julia/packages/Plots/Xv2qA/src/args.jl:1575
┌ Warning: Keyword argument hover not supported with Plots.PyPlotBackend().  Choose from: Set([:zforeground_color_guide, :top_margin, :group, :inset_subplots, :stride, :background_color, :ytickfontsize, :yforeground_color_text, :yguidefontcolor, :tickfontfamily, :show_empty_bins, :seriesalpha, :seriescolor, :xrotation, :ztick_direction, :xgrid, :colorbar_tickfontfamily, :ygridalpha, :colorbar_tickfonthalign, :dpi, :zlims, :colorbar, :xtick_direction, :legend_font_family, :zflip, :ticks, :linealpha, :overwrite_figure, :arrow, :normalize, :linestyle, :xflip, :yforeground_color_guide, :zgrid, :fillcolor, :ygrid, :bar_width, :colorbar_scale, :background_color_inside, :bins, :zguide, :zforeground_color_text, :yscale, :legend_font_color, :weights, :xgridalpha, :ygridstyle, :clims, :xtickfontcolor, :fill_z, :xguide, :markershape, :background_color_subplot, :colorbar_tickfontrotation, :fillalpha, :markerstrokewidth, :tick_direction, :ztickfontfamily, :xguidefontfamily, :gridlinewidth, :foreground_color_subplot, :xgridlinewidth, :yguidefontsize, :foreground_color, :foreground_color_text, :x, :yerror, :rotation, :zgridlinewidth, :discrete_values, :ytick_direction, :grid, :ribbon, :xguidefontsize, :xforeground_color_axis, :xdiscrete_values, :background_color_outside, :titlefontcolor, :xgridstyle, :line_z, :size, :orientation, :gridstyle, :projection, :markersize, :legend_foreground_color, :camera, :ydiscrete_values, :xforeground_color_grid, :seriestype, :yflip, :quiver, :zticks, :markerstrokecolor, :fillrange, :colorbar_ticks, :xlims, :xforeground_color_border, :markercolor, :yrotation, :xtickfontsize, :ylink, :levels, :color_palette, :connections, :yforeground_color_grid, :lims, :zgridstyle, :foreground_color_border, :xscale, :marker_z, :markerstrokealpha, :left_margin, :markeralpha, :annotations, :window_title, :zrotation, :foreground_color_axis, :zguidefontcolor, :ygridlinewidth, :zlink, :zscale, :smooth, :colorbar_tickfontcolor, :xticks, :guidefontsize, :zguidefontsize, :y, :margin, :ytickfontcolor, :foreground_color_title, :bottom_margin, :zdiscrete_values, :yforeground_color_border, :zguidefontfamily, :framestyle, :zgridalpha, :zerror, :ztickfontcolor, :scale, :legend_position, :linecolor, :html_output_format, :legend_title, :zforeground_color_border, :yguide_position, :legend_font_pointsize, :contours, :title, :tickfontcolor, :subplot_index, :flip, :xguide_position, :legend_background_color, :tickfontsize, :z, :yforeground_color_axis, :foreground_color_grid, :linewidth, :ztickfontsize, :gridalpha, :xerror, :guidefontfamily, :ylims, :bar_position, :contour_labels, :zguide_position, :xguidefontcolor, :primary, :colorbar_tickfontsize, :xtickfontfamily, :ytickfontfamily, :aspect_ratio, :xforeground_color_text, :foreground_color_guide, :show, :link, :guide_position, :colorbar_title, :xforeground_color_guide, :subplot, :label, :guide, :titlelocation, :guidefontcolor, :colorbar_tickfontvalign, :bar_edges, :yguide, :titlefontsize, :titlefont, :titlefontfamily, :zforeground_color_axis, :yguidefontfamily, :zforeground_color_grid, :layout, :colorbar_entry, :polar, :right_margin, :xlink, :series_annotations, :yticks])
└ @ Plots ~/.julia/packages/Plots/Xv2qA/src/args.jl:1607</code></pre><p><img src="../he-plot.svg" alt/></p><h3 id="Filtering"><a class="docs-heading-anchor" href="#Filtering">Filtering</a><a id="Filtering-1"></a><a class="docs-heading-anchor-permalink" href="#Filtering" title="Permalink"></a></h3><p>Two filtering function are available for filtering cells and features/genes based on basic criteria such as the number of features detected, the number of cells for which a feature is detected and the total number of counts.</p><pre><code class="language-julia hljs">X = filter_cells(X, min_features=200)
X = filter_features(X, min_cells=3)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">2700×13714 Named sparse matrix with 2282976 Int32 nonzero entries:
	[AGATTCCTGACGAG-1,          &quot;AL627309.1&quot;]  =  1
	[CCAAAGTGCTACGA-1,          &quot;AL627309.1&quot;]  =  1
	[CCGTACACGTTGGT-1,          &quot;AL627309.1&quot;]  =  1
	[CGACTGCTTCCTCG-1,          &quot;AL627309.1&quot;]  =  1
	[CTATGTTGTCTCGC-1,          &quot;AL627309.1&quot;]  =  1
	[CTGTGAGACGAACT-1,          &quot;AL627309.1&quot;]  =  1
	[GATACTCTATCGGT-1,          &quot;AL627309.1&quot;]  =  1
	[GCTAGATGAGCTCA-1,          &quot;AL627309.1&quot;]  =  1
	[GCTTAACTACTGGT-1,          &quot;AL627309.1&quot;]  =  1
	                                           ⋮
	[TCTCAAACCTAAGC-1,            &quot;SRSF10.1&quot;]  =  1
	[TGAGACACTCAAGC-1,            &quot;SRSF10.1&quot;]  =  1
	[TGATCGGATATGCG-1,            &quot;SRSF10.1&quot;]  =  1
	[TGTAGGTGCTATGG-1,            &quot;SRSF10.1&quot;]  =  1
	[TTATCCGAGAAAGT-1,            &quot;SRSF10.1&quot;]  =  1
	[TTCAGTACCGACTA-1,            &quot;SRSF10.1&quot;]  =  1
	[TTGAATGATCTCAT-1,            &quot;SRSF10.1&quot;]  =  1
	[TTGAGGACTACGCA-1,            &quot;SRSF10.1&quot;]  =  1
	[TTTATCCTGTTGTG-1,            &quot;SRSF10.1&quot;]  =  1
	[TTTCCAGAGGTGAG-1,            &quot;SRSF10.1&quot;]  =  1</code></pre><p>This filters out cells with less than 200 features/genes and features detected in less than 3 cells. For convenience, <code>filter_counts</code> combines these two into a single function, similar to Seurat. Beware that the order in which you call these functions can matter.</p><article class="docstring"><header><a class="docstring-binding" id="Severo.filter_features" href="#Severo.filter_features"><code>Severo.filter_features</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">filter_features(; min_cells = 0)</code></pre><p>Partial version of filter<em>features(A::NamedCountMatrix; min</em>cells = 0)</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/862003e8b61d12b112e2e343b7f508c1a959ad97/src/macros.jl#L81-L85">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Severo.filter_cells" href="#Severo.filter_cells"><code>Severo.filter_cells</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">filter_cells(; min_features = 0, min_feature_count = 0, min_umi = 0)</code></pre><p>Partial version of filter<em>cells(A::NamedCountMatrix; min</em>features = 0, min<em>feature</em>count = 0, min_umi = 0)</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/862003e8b61d12b112e2e343b7f508c1a959ad97/src/macros.jl#L81-L85">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Severo.filter_counts" href="#Severo.filter_counts"><code>Severo.filter_counts</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">filter_counts(; min_cells = 0, min_features = 0, min_feature_count = 0, min_umi = 0)</code></pre><p>Partial version of filter<em>counts(A::NamedCountMatrix; min</em>cells = 0, min<em>features = 0, min</em>feature<em>count = 0, min</em>umi = 0)</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/862003e8b61d12b112e2e343b7f508c1a959ad97/src/macros.jl#L81-L85">source</a></section></article><h3 id="Normalization"><a class="docs-heading-anchor" href="#Normalization">Normalization</a><a id="Normalization-1"></a><a class="docs-heading-anchor-permalink" href="#Normalization" title="Permalink"></a></h3><p>After removing unwanted cells from the dataset, the next step is to normalize the data. The basic method normalizes the feature expression measurements for each cell by the total expression, multiplies this by a scale factor and optionally log-transforming the result.</p><pre><code class="language-julia hljs">Y = normalize_cells(X, method=:lognormalize, scale_factor=1e4)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">2700×13714 Named sparse matrix with 2282976 Float64 nonzero entries:
	[AGATTCCTGACGAG-1,          &quot;AL627309.1&quot;]  =  1.538222672557476
	[CCAAAGTGCTACGA-1,          &quot;AL627309.1&quot;]  =  1.3631966075373008
	[CCGTACACGTTGGT-1,          &quot;AL627309.1&quot;]  =  1.3945254817849748
	[CGACTGCTTCCTCG-1,          &quot;AL627309.1&quot;]  =  1.787190785781838
	[CTATGTTGTCTCGC-1,          &quot;AL627309.1&quot;]  =  1.3514100666239552
	[CTGTGAGACGAACT-1,          &quot;AL627309.1&quot;]  =  1.6848688061171004
	[GATACTCTATCGGT-1,          &quot;AL627309.1&quot;]  =  1.6623488639622728
	[GCTAGATGAGCTCA-1,          &quot;AL627309.1&quot;]  =  1.6749250073944073
	[GCTTAACTACTGGT-1,          &quot;AL627309.1&quot;]  =  1.8821985210730219
	                                           ⋮
	[TCTCAAACCTAAGC-1,            &quot;SRSF10.1&quot;]  =  2.050332029149289
	[TGAGACACTCAAGC-1,            &quot;SRSF10.1&quot;]  =  1.873284442922057
	[TGATCGGATATGCG-1,            &quot;SRSF10.1&quot;]  =  2.021955420152209
	[TGTAGGTGCTATGG-1,            &quot;SRSF10.1&quot;]  =  1.2128182558075358
	[TTATCCGAGAAAGT-1,            &quot;SRSF10.1&quot;]  =  1.2133165770732028
	[TTCAGTACCGACTA-1,            &quot;SRSF10.1&quot;]  =  1.529100016033425
	[TTGAATGATCTCAT-1,            &quot;SRSF10.1&quot;]  =  1.7863632500815836
	[TTGAGGACTACGCA-1,            &quot;SRSF10.1&quot;]  =  0.9465443048463601
	[TTTATCCTGTTGTG-1,            &quot;SRSF10.1&quot;]  =  1.31322083398467
	[TTTCCAGAGGTGAG-1,            &quot;SRSF10.1&quot;]  =  1.7178390681467346</code></pre><article class="docstring"><header><a class="docstring-binding" id="Severo.normalize_cells" href="#Severo.normalize_cells"><code>Severo.normalize_cells</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">normalize_cells(; method = :lognormalize, scale_factor::Real = 1.0, dtype::Type{T} = Float64) where T &lt;: AbstractFloat</code></pre><p>Partial version of normalize<em>cells(X::NamedCountMatrix; method = :lognormalize, scale</em>factor::Real = 1.0, dtype::Type{T} = Float64) where T &lt;: AbstractFloat</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/862003e8b61d12b112e2e343b7f508c1a959ad97/src/macros.jl#L81-L85">source</a></section></article><h3 id="Identifying-highly-variable-features"><a class="docs-heading-anchor" href="#Identifying-highly-variable-features">Identifying highly variable features</a><a id="Identifying-highly-variable-features-1"></a><a class="docs-heading-anchor-permalink" href="#Identifying-highly-variable-features" title="Permalink"></a></h3><p>Next, we select a subset of features that exhibit high cell-to-cell variation in the dataset (i.e, they are highly expressed in some cells, and lowly expressed in others).</p><pre><code class="language-julia hljs">hvf = find_variable_features(X, 2000; method=:vst)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">2000-element Named SubArray{Int64, 1, Vector{Int64}, Tuple{UnitRange{Int64}}, true}
features              │ 
──────────────────────┼──────
&quot;PPBP&quot;                │  3316
&quot;LYZ&quot;                 │  8682
&quot;S100A9&quot;              │   908
&quot;IGLL5&quot;               │ 13275
&quot;GNLY&quot;                │  1804
&quot;FTL&quot;                 │ 12944
&quot;PF4&quot;                 │  3315
&quot;FTH1&quot;                │  7755
⋮                           ⋮
&quot;EIF3D&quot;               │ 13383
&quot;POLR2I&quot;              │ 12681
&quot;ZCCHC10&quot;             │  3924
&quot;SEC61B&quot;              │  6669
&quot;RP11-164H13.1&quot;       │  9585
&quot;NELL2&quot;               │  8455
&quot;NF1&quot;                 │ 11047
&quot;GTF2H2&quot;              │  3747</code></pre><article class="docstring"><header><a class="docstring-binding" id="Severo.find_variable_features" href="#Severo.find_variable_features"><code>Severo.find_variable_features</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">find_variable_features(nfeatures = 2000; method = :vst, dtype::Type{&lt;:AbstractFloat} = Float64, kw...)</code></pre><p>Partial version of find<em>variable</em>features(counts::NamedCountMatrix, nfeatures = 2000; method = :vst, dtype::Type{&lt;:AbstractFloat} = Float64, kw...)</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/862003e8b61d12b112e2e343b7f508c1a959ad97/src/macros.jl#L81-L85">source</a></section></article><h2 id="Dimensionality-reduction"><a class="docs-heading-anchor" href="#Dimensionality-reduction">Dimensionality reduction</a><a id="Dimensionality-reduction-1"></a><a class="docs-heading-anchor-permalink" href="#Dimensionality-reduction" title="Permalink"></a></h2><h3 id="Scaling"><a class="docs-heading-anchor" href="#Scaling">Scaling</a><a id="Scaling-1"></a><a class="docs-heading-anchor-permalink" href="#Scaling" title="Permalink"></a></h3><p>Prior to dimensional reduction techniques like PCA, it&#39;s a good idea to <code>scale_features</code> the data. Scaling performs two basic operations:</p><ul><li>Shifts the expression of each gene, so that the mean expression across cells is 0</li><li>Scales the expression of each gene, so that the standard deviation across cells is 1. This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate</li><li>Clips values exceeding standard deviation of <code>scale_max</code></li></ul><pre><code class="language-julia hljs">Y = Y[:,hvf] # only use highly-variable features
S = scale_features(Y; scale_max=10)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">CenteredMatrix:
  A = 2700×2000 Named sparse matrix with 515841 Float64 nonzero entries:
	[AAACCGTGCTTCCG-1,                &quot;PPBP&quot;]  =  3.05759
	[AAATCAACCCTATT-1,                &quot;PPBP&quot;]  =  1.983
	[AAATCCCTGCTATG-1,                &quot;PPBP&quot;]  =  5.03146
	[AACCGCCTCTACGA-1,                &quot;PPBP&quot;]  =  2.44748
	[AAGCCTGACATGCA-1,                &quot;PPBP&quot;]  =  3.68051
	[AATCCTTGACGGGA-1,                &quot;PPBP&quot;]  =  3.12117
	[ACCCACTGGTTCAG-1,                &quot;PPBP&quot;]  =  10.143
	[ACCTGAGATATCGG-1,                &quot;PPBP&quot;]  =  9.63048
	[ACCTGGCTGTCTTT-1,                &quot;PPBP&quot;]  =  7.59301
	                                           ⋮
	[TGAGTGACTGAGCT-1,              &quot;GTF2H2&quot;]  =  7.15279
	[TGCGATGACCTCGT-1,              &quot;GTF2H2&quot;]  =  7.51358
	[TGGGTATGAAGAGT-1,              &quot;GTF2H2&quot;]  =  6.33385
	[TGTGAGACTGTCAG-1,              &quot;GTF2H2&quot;]  =  6.86631
	[TTAGGGACGCGAAG-1,              &quot;GTF2H2&quot;]  =  7.54787
	[TTAGTCTGAAAGCA-1,              &quot;GTF2H2&quot;]  =  6.76249
	[TTCCAAACTCCCAC-1,              &quot;GTF2H2&quot;]  =  10.1315
	[TTCGTATGTCCTTA-1,              &quot;GTF2H2&quot;]  =  5.26091
	[TTGGTACTGAATCC-1,              &quot;GTF2H2&quot;]  =  8.17128
	[TTTAGCTGTACTCT-1,              &quot;GTF2H2&quot;]  =  4.47778
  mu = 2000-element Named Vector{Float64}
features              │ 
──────────────────────┼──────────
&quot;PPBP&quot;                │  0.142967
&quot;LYZ&quot;                 │  0.972527
&quot;S100A9&quot;              │  0.645189
&quot;IGLL5&quot;               │   0.18421
&quot;GNLY&quot;                │  0.407293
&quot;FTL&quot;                 │   2.94261
&quot;PF4&quot;                 │  0.110369
&quot;FTH1&quot;                │   3.22873
⋮                               ⋮
&quot;EIF3D&quot;               │  0.775623
&quot;POLR2I&quot;              │  0.396716
&quot;ZCCHC10&quot;             │  0.211964
&quot;SEC61B&quot;              │  0.656028
&quot;RP11-164H13.1&quot;       │ 0.0889054
&quot;NELL2&quot;               │   0.25572
&quot;NF1&quot;                 │  0.130034
&quot;GTF2H2&quot;              │  0.131466</code></pre><article class="docstring"><header><a class="docstring-binding" id="Severo.scale_features" href="#Severo.scale_features"><code>Severo.scale_features</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">scale_features(; scale_max::Real = Inf, dtype::Type{&lt;:AbstractFloat} = Float64, features::Union{Nothing, AbstractArray} = nothing)</code></pre><p>Partial version of scale<em>features(X::(NamedArray{T, 2, SparseMatrixCSC{T, Int64}} where T); scale</em>max::Real = Inf, dtype::Type{&lt;:AbstractFloat} = Float64, features::Union{Nothing, AbstractArray} = nothing)</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/862003e8b61d12b112e2e343b7f508c1a959ad97/src/macros.jl#L81-L85">source</a></section></article><h3 id="Principal-component-analysis"><a class="docs-heading-anchor" href="#Principal-component-analysis">Principal component analysis</a><a id="Principal-component-analysis-1"></a><a class="docs-heading-anchor-permalink" href="#Principal-component-analysis" title="Permalink"></a></h3><p>Reduce the dimensionality of the data by running principal component analysis (PCA), which reveals the main axes of variation and de-noises the data.</p><pre><code class="language-julia hljs">em = embedding(S, 15, method=:pca)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">PC-1
Positive: InlineStrings.String31[&quot;CST3&quot;, &quot;TYROBP&quot;, &quot;LST1&quot;, &quot;AIF1&quot;, &quot;FTL&quot;, &quot;FCN1&quot;, &quot;LYZ&quot;, &quot;FTH1&quot;, &quot;FCER1G&quot;, &quot;S100A9&quot;]
Negative: InlineStrings.String31[&quot;LTB&quot;, &quot;IL32&quot;, &quot;CD2&quot;, &quot;ACAP1&quot;, &quot;STK17A&quot;, &quot;CTSW&quot;, &quot;CD247&quot;, &quot;GIMAP5&quot;, &quot;AQP3&quot;, &quot;CCL5&quot;]
PC-2
Positive: InlineStrings.String31[&quot;CD79A&quot;, &quot;MS4A1&quot;, &quot;TCL1A&quot;, &quot;HLA-DRA&quot;, &quot;HLA-DQA1&quot;, &quot;HLA-DQB1&quot;, &quot;LINC00926&quot;, &quot;CD79B&quot;, &quot;HLA-DRB1&quot;, &quot;CD74&quot;]
Negative: InlineStrings.String31[&quot;NKG7&quot;, &quot;PRF1&quot;, &quot;CST7&quot;, &quot;GZMA&quot;, &quot;GZMB&quot;, &quot;FGFBP2&quot;, &quot;CTSW&quot;, &quot;GNLY&quot;, &quot;GZMH&quot;, &quot;SPON2&quot;]
PC-3
Positive: InlineStrings.String31[&quot;HLA-DQA1&quot;, &quot;CD79A&quot;, &quot;HLA-DQB1&quot;, &quot;CD79B&quot;, &quot;HLA-DPB1&quot;, &quot;CD74&quot;, &quot;HLA-DPA1&quot;, &quot;MS4A1&quot;, &quot;HLA-DRB1&quot;, &quot;HLA-DRB5&quot;]
Negative: InlineStrings.String31[&quot;PPBP&quot;, &quot;PF4&quot;, &quot;SDPR&quot;, &quot;SPARC&quot;, &quot;GNG11&quot;, &quot;NRGN&quot;, &quot;GP9&quot;, &quot;RGS18&quot;, &quot;TUBB1&quot;, &quot;HIST1H2AC&quot;]
PC-4
Positive: InlineStrings.String31[&quot;HLA-DQA1&quot;, &quot;CD79A&quot;, &quot;CD79B&quot;, &quot;HLA-DQB1&quot;, &quot;MS4A1&quot;, &quot;CD74&quot;, &quot;HLA-DPB1&quot;, &quot;HIST1H2AC&quot;, &quot;PF4&quot;, &quot;HLA-DQA2&quot;]
Negative: InlineStrings.String31[&quot;RPS2&quot;, &quot;VIM&quot;, &quot;S100A6&quot;, &quot;S100A8&quot;, &quot;S100A4&quot;, &quot;IL32&quot;, &quot;S100A9&quot;, &quot;TMSB10&quot;, &quot;GIMAP7&quot;, &quot;S100A10&quot;]
PC-5
Positive: InlineStrings.String31[&quot;GZMB&quot;, &quot;FGFBP2&quot;, &quot;NKG7&quot;, &quot;GNLY&quot;, &quot;PRF1&quot;, &quot;CCL4&quot;, &quot;CST7&quot;, &quot;SPON2&quot;, &quot;GZMA&quot;, &quot;GZMH&quot;]
Negative: InlineStrings.String31[&quot;LTB&quot;, &quot;VIM&quot;, &quot;AQP3&quot;, &quot;PPA1&quot;, &quot;MAL&quot;, &quot;RPS2&quot;, &quot;KIAA0101&quot;, &quot;CD2&quot;, &quot;CORO1B&quot;, &quot;CYTIP&quot;]
</code></pre><pre><code class="language-julia hljs">plot_elbow(em)</code></pre><p><img src="../elbow-plot.svg" alt/></p><article class="docstring"><header><a class="docstring-binding" id="Severo.embedding" href="#Severo.embedding"><code>Severo.embedding</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">embedding(ncomponents::Int64 = 50; method = :pca, kw...)</code></pre><p>Partial version of embedding(X, ncomponents::Int64 = 50; method = :pca, kw...)</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/862003e8b61d12b112e2e343b7f508c1a959ad97/src/macros.jl#L81-L85">source</a></section></article><h2 id="Clustering-the-cells"><a class="docs-heading-anchor" href="#Clustering-the-cells">Clustering the cells</a><a id="Clustering-the-cells-1"></a><a class="docs-heading-anchor-permalink" href="#Clustering-the-cells" title="Permalink"></a></h2><p>Briefly, the clustering embed cells in a graph structure with edges between cells with similar feature expression patterns, and then attempt to partition this graph into highly interconnected &#39;quasi-cliques&#39; or &#39;communities&#39;.</p><h3 id="Computing-the-neighborhood-graph"><a class="docs-heading-anchor" href="#Computing-the-neighborhood-graph">Computing the neighborhood graph</a><a id="Computing-the-neighborhood-graph-1"></a><a class="docs-heading-anchor-permalink" href="#Computing-the-neighborhood-graph" title="Permalink"></a></h3><p>First, we need to find the k-nearest neighbors of all the cells based on the euclidean distance in PCA space, and then compute a graph based on the shared overlap in the local neighborhoods between any two cells (Jaccard similarity). These step are performed using the <code>nearest_neighbours</code> and <code>jaccard_index</code> functions. For convenience, the two steps are also combined in the <code>shared_nearest_neighours</code> function.</p><pre><code class="language-julia hljs">nn = nearest_neighbours(em, 20, dims=1:10)
snn = jaccard_index(nn, prune=1/15)
# or as a single step
#snn = nearest_neighbours(em, 20, dims=1:10, prune=1/15)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">2700×2700 Named sparse matrix with 198538 Float64 nonzero entries:
	[AAACATACAACCAC-1, AAACATACAACCAC-1]  =  1.0
	[AAATTGACTCGCTC-1, AAACATACAACCAC-1]  =  0.08108108108108109
	[AACTCTTGCAGGAG-1, AAACATACAACCAC-1]  =  0.1111111111111111
	[AATCTAGAATCGGT-1, AAACATACAACCAC-1]  =  0.08108108108108109
	[ACCACGCTGCTGTA-1, AAACATACAACCAC-1]  =  0.14285714285714285
	[ACGCCGGAAATGCC-1, AAACATACAACCAC-1]  =  0.14285714285714285
	[ACTAGGTGGAACTC-1, AAACATACAACCAC-1]  =  0.08108108108108109
	[AGAGATGATTGTGG-1, AAACATACAACCAC-1]  =  0.08108108108108109
	[AGCCGTCTGAGAGC-1, AAACATACAACCAC-1]  =  0.08108108108108109
	                                      ⋮
	[TCGGTAGAGTAGGG-1, TTTGCATGCCTCAC-1]  =  0.08108108108108109
	[TGATAAACTTTCAC-1, TTTGCATGCCTCAC-1]  =  0.08108108108108109
	[TGATACCTGTTGGT-1, TTTGCATGCCTCAC-1]  =  0.08108108108108109
	[TGCCAGCTTGGCAT-1, TTTGCATGCCTCAC-1]  =  0.14285714285714285
	[TGGTTACTGTTCTT-1, TTTGCATGCCTCAC-1]  =  0.1111111111111111
	[TGTAACCTAGAGGC-1, TTTGCATGCCTCAC-1]  =  0.1111111111111111
	[TTAGAATGTGGTGT-1, TTTGCATGCCTCAC-1]  =  0.08108108108108109
	[TTAGCTACAACCGT-1, TTTGCATGCCTCAC-1]  =  0.1111111111111111
	[TTCGATTGAGCATC-1, TTTGCATGCCTCAC-1]  =  0.14285714285714285
	[TTTGCATGCCTCAC-1, TTTGCATGCCTCAC-1]  =  1.0</code></pre><article class="docstring"><header><a class="docstring-binding" id="Severo.nearest_neighbours" href="#Severo.nearest_neighbours"><code>Severo.nearest_neighbours</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">nearest_neighbours(k::Int64; dims = (:), metric::SemiMetric = Euclidean(), include_self::Bool = true, ntables::Int64 = 2 * size(X, 2)) where T</code></pre><p>Partial version of nearest<em>neighbours(X::NamedArray{T, 2}, k::Int64; dims = (:), metric::SemiMetric = Euclidean(), include</em>self::Bool = true, ntables::Int64 = 2 * size(X, 2)) where T</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/862003e8b61d12b112e2e343b7f508c1a959ad97/src/macros.jl#L81-L85">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Severo.jaccard_index" href="#Severo.jaccard_index"><code>Severo.jaccard_index</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">jaccard_index(k::Int64; prune::Real = 1 / 15, dtype::Type{R} = Float64) where R &lt;: AbstractFloat</code></pre><p>Partial version of jaccard_index(nn::(NamedArray{T, 2} where T), k::Int64; prune::Real = 1 / 15, dtype::Type{R} = Float64) where R &lt;: AbstractFloat</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/862003e8b61d12b112e2e343b7f508c1a959ad97/src/macros.jl#L81-L85">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Severo.shared_nearest_neighbours" href="#Severo.shared_nearest_neighbours"><code>Severo.shared_nearest_neighbours</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">shared_nearest_neighbours(k::Int64; kw...) where T</code></pre><p>Partial version of shared<em>nearest</em>neighbours(X::NamedArray{T, 2}, k::Int64; kw...) where T</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/862003e8b61d12b112e2e343b7f508c1a959ad97/src/macros.jl#L81-L85">source</a></section></article><h3 id="Clustering-the-neighborhood-graph"><a class="docs-heading-anchor" href="#Clustering-the-neighborhood-graph">Clustering the neighborhood graph</a><a id="Clustering-the-neighborhood-graph-1"></a><a class="docs-heading-anchor-permalink" href="#Clustering-the-neighborhood-graph" title="Permalink"></a></h3><p>Clustering is performed using a graph-based method called <code>Louvain clustering</code> which tries to detect communities based on optimizing a modularity function. Since this is a greedy algorithm, which can get stuck in local minima based on random initialization, multiple restarts are required to find a more &quot;global&quot; minima.</p><pre><code class="language-julia hljs">lbls = cluster(snn)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">2700-element Named Vector{Int64}
cells            │ 
─────────────────┼───
AAACATACAACCAC-1 │  1
AAACATTGAGCTAC-1 │  2
AAACATTGATCAGC-1 │  1
AAACCGTGCTTCCG-1 │  3
AAACCGTGTATGCG-1 │  4
AAACGCACTGGTAC-1 │  1
AAACGCTGACCAGT-1 │  5
AAACGCTGGTTCTT-1 │  5
⋮                   ⋮
TTTCAGTGTGCAGT-1 │  2
TTTCCAGAGGTGAG-1 │  1
TTTCGAACACCTGA-1 │ 10
TTTCGAACTCTCAT-1 │  8
TTTCTACTGAGGCA-1 │  2
TTTCTACTTCCTCG-1 │  2
TTTGCATGAGAGGC-1 │  2
TTTGCATGCCTCAC-1 │  7</code></pre><article class="docstring"><header><a class="docstring-binding" id="Severo.cluster" href="#Severo.cluster"><code>Severo.cluster</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">cluster(; kw...)</code></pre><p>Partial version of cluster(SNN::NeighbourGraph; kw...)</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/862003e8b61d12b112e2e343b7f508c1a959ad97/src/macros.jl#L81-L85">source</a></section></article><h2 id="Finding-differentially-expressed-features"><a class="docs-heading-anchor" href="#Finding-differentially-expressed-features">Finding differentially expressed features</a><a id="Finding-differentially-expressed-features-1"></a><a class="docs-heading-anchor-permalink" href="#Finding-differentially-expressed-features" title="Permalink"></a></h2><p>We can find markers that define clusters via differential expression. It identifies positive and negative markers for a single cluster compared to the other cells. <code>find_markers</code> performs this process for every cluster. The function returns a <code>DataFrame</code> containing for each marker: the p-value, the log-foldchange and statistical score.</p><pre><code class="language-julia hljs">dx = find_markers(X, lbls; method=:wilcoxon)
dx = filter_rank_markers(dx)
dx[1:10, :]</code></pre><div class="data-frame"><p>10 rows × 5 columns</p><table class="data-frame"><thead><tr><th></th><th>group</th><th>score</th><th>pval</th><th>logfc</th><th>feature</th></tr><tr><th></th><th title="Int64">Int64</th><th title="Float64">Float64</th><th title="Float64">Float64</th><th title="Float64">Float64</th><th title="InlineStrings.String31">String31</th></tr></thead><tbody><tr><th>1</th><td>1</td><td>15.5809</td><td>9.8217e-55</td><td>-1.67251</td><td>HLA-DRB1</td></tr><tr><th>2</th><td>1</td><td>15.2618</td><td>1.37483e-52</td><td>-2.51018</td><td>HLA-DRA</td></tr><tr><th>3</th><td>1</td><td>15.1851</td><td>4.43631e-52</td><td>-2.2133</td><td>TYROBP</td></tr><tr><th>4</th><td>1</td><td>14.5636</td><td>4.78877e-48</td><td>-1.66148</td><td>FCER1G</td></tr><tr><th>5</th><td>1</td><td>14.4178</td><td>3.99753e-47</td><td>-1.06897</td><td>HLA-DRB5</td></tr><tr><th>6</th><td>1</td><td>14.224</td><td>6.50439e-46</td><td>-1.64141</td><td>HLA-DPA1</td></tr><tr><th>7</th><td>1</td><td>13.7386</td><td>5.96083e-43</td><td>-1.64782</td><td>HLA-DPB1</td></tr><tr><th>8</th><td>1</td><td>12.4899</td><td>8.47673e-36</td><td>-1.94933</td><td>CD74</td></tr><tr><th>9</th><td>1</td><td>12.3375</td><td>5.68872e-35</td><td>-0.694127</td><td>HLA-DMA</td></tr><tr><th>10</th><td>1</td><td>12.2333</td><td>2.06408e-34</td><td>-0.850923</td><td>CFD</td></tr></tbody></table></div><p>Performance can be improved by first filtering the features based on the percentage of cells and the fold-change between the groups.</p><pre><code class="language-julia hljs"># filtering and DE on the lognormalized data
sel = prefilter_markers(Y, lbls, logfc_threshold=0.25, min_pct=0.1, log=true, only_pos=true)
dx = find_markers(Y, lbls; method=:t, log=true, selection=sel) # t-test
dx = filter_rank_markers(dx)
dx[1:10, :]</code></pre><div class="data-frame"><p>10 rows × 5 columns</p><table class="data-frame"><thead><tr><th></th><th>group</th><th>score</th><th>pval</th><th>logfc</th><th>feature</th></tr><tr><th></th><th title="Int64">Int64</th><th title="Float64">Float64</th><th title="Float64">Float64</th><th title="Float64">Float64</th><th title="String">String</th></tr></thead><tbody><tr><th>1</th><td>1</td><td>30.6481</td><td>2.0335e-152</td><td>1.22346</td><td>IL32</td></tr><tr><th>2</th><td>1</td><td>29.1476</td><td>5.85058e-143</td><td>1.28312</td><td>LTB</td></tr><tr><th>3</th><td>1</td><td>15.4274</td><td>2.16513e-46</td><td>1.21357</td><td>CD2</td></tr><tr><th>4</th><td>1</td><td>14.4696</td><td>1.37475e-43</td><td>0.490875</td><td>VIM</td></tr><tr><th>5</th><td>1</td><td>14.0062</td><td>3.44367e-40</td><td>0.582623</td><td>GIMAP7</td></tr><tr><th>6</th><td>1</td><td>12.7467</td><td>3.65549e-33</td><td>1.23198</td><td>AQP3</td></tr><tr><th>7</th><td>1</td><td>11.4712</td><td>2.27143e-28</td><td>0.599281</td><td>ANXA1</td></tr><tr><th>8</th><td>1</td><td>10.6418</td><td>1.9189e-24</td><td>0.853336</td><td>TRADD</td></tr><tr><th>9</th><td>1</td><td>9.21204</td><td>3.80925e-19</td><td>0.769177</td><td>TNFAIP8</td></tr><tr><th>10</th><td>1</td><td>8.67919</td><td>3.45651e-17</td><td>0.716185</td><td>MAL</td></tr></tbody></table></div><article class="docstring"><header><a class="docstring-binding" id="Severo.find_markers" href="#Severo.find_markers"><code>Severo.find_markers</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">find_markers(X::Union{NamedCountMatrix, NamedDataMatrix}, idents::NamedVector{&lt;:Integer};
    method=:wilcoxon, selection::Union{Nothing, NamedArray{Bool, 2}, AbstractArray{Bool,2}}=nothing, log::Bool=false, kw...)</code></pre><p>Finds markers (differentially expressed genes) for each of the classes in a dataset.</p><p><strong>Arguments</strong>:</p><pre><code class="nohighlight hljs">-`X`: count or data matrix
-`idents`: class identity for each cell
-`method`: Which test to use, supported are: [wilcoxon, t]
-`selection`: a selection of features and groups that should be considered
-`log`: the data is in log-scale (default = false)
-`kw...`: additional parameters passed down to the method</code></pre><p><strong>Return values</strong>:</p><p>A <code>DataFrame</code> containing a list of putative markers with associated statistics (p-values and scores) and log fold-changes.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/862003e8b61d12b112e2e343b7f508c1a959ad97/src/diffexpr.jl#L192-L210">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Severo.filter_rank_markers" href="#Severo.filter_rank_markers"><code>Severo.filter_rank_markers</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">filter_rank_markers(de::DataFrame; pval_thresh::Real=1e-2, ngenes::Integer=typemax(Int64))</code></pre><p>Filters and ranks a list of markers (differentially expressed genes).</p><p><strong>Arguments</strong>:</p><pre><code class="nohighlight hljs">-`de`: list of markers returned by [find_markers](@ref)
-`pval_thresh`: only keep markers with pval &lt; pval_thresh
-`count`: the number of highest-ranked markers to keep
-`rankby_abs`: rank based on absolute value of the scores</code></pre><p><strong>Return values</strong>:</p><p>A <code>DataFrame</code> containing a ranked list of filtered markers.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/862003e8b61d12b112e2e343b7f508c1a959ad97/src/diffexpr.jl#L320-L335">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Severo.prefilter_markers" href="#Severo.prefilter_markers"><code>Severo.prefilter_markers</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">prefilter_markers(X::Union{NamedCountMatrix, NamedDataMatrix}, idents::NamedVector{&lt;:Integer};
        logfc_threshold::Real=0.0, min_pct::Real=0.0, min_diff_pct::Real=-Inf, only_pos:Bool=false, log::Bool=false)</code></pre><p>Filter features for each of the classes in a dataset.</p><p><strong>Arguments</strong>:</p><pre><code class="nohighlight hljs">-`X`: count or data matrix
-`idents`: class identity for each cell
-`logfc_threshold`: Limit testing to features which show, on average, at least X-fold difference (log-scale) between the two groups of cells
-`min_pct`: only test features that are detected in a minimum fraction of `min_pct` cells in either of the two populations
-`min_diff_pct`: only test features that show a minimum difference in the fraction of detection between the two groups.
-`only_pos`: only return features with positive log fold-change
-`log`: the data is in log-scale (default = false)</code></pre><p><strong>Return values</strong>:</p><p>Selection matrix for each feature and class</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/862003e8b61d12b112e2e343b7f508c1a959ad97/src/diffexpr.jl#L121-L140">source</a></section></article><pre><code class="language- hljs">plot_violin(X, [&quot;CST3&quot;, &quot;NKG7&quot;, &quot;PPBP&quot;, &quot;S100A4&quot;], lbls)
savefig(&quot;violin-plot.svg&quot;); nothing # hide</code></pre><p><img src="violin-plot.svg" alt/></p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../">« Home</a><a class="docs-footer-nextpage" href="../performance/">Performance »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.12 on <span class="colophon-date" title="Wednesday 2 February 2022 09:05">Wednesday 2 February 2022</span>. Using Julia version 1.6.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
