<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Preprocessing and clustering 3k PBMCs · Severo.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">Severo.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><span class="tocitem">Showcases</span><ul><li class="is-active"><a class="tocitem" href>Preprocessing and clustering 3k PBMCs</a><ul class="internal"><li><a class="tocitem" href="#Overview"><span>Overview</span></a></li><li><a class="tocitem" href="#Loading-data"><span>Loading data</span></a></li><li><a class="tocitem" href="#Preprocessing"><span>Preprocessing</span></a></li><li><a class="tocitem" href="#Dimensionality-reduction"><span>Dimensionality reduction</span></a></li><li><a class="tocitem" href="#Clustering-the-cells"><span>Clustering the cells</span></a></li><li><a class="tocitem" href="#Finding-differentially-expressed-features"><span>Finding differentially expressed features</span></a></li></ul></li><li><a class="tocitem" href="../performance/">Performance</a></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../public/">Public</a></li><li><a class="tocitem" href="../internals/">Internals</a></li></ul></li><li><a class="tocitem" href="../contributing/">Contributing</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Showcases</a></li><li class="is-active"><a href>Preprocessing and clustering 3k PBMCs</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Preprocessing and clustering 3k PBMCs</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/ExaScience/Severo.jl/blob/master/docs/src/pbmc.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Preprocessing-and-clustering-3k-PBMCs"><a class="docs-heading-anchor" href="#Preprocessing-and-clustering-3k-PBMCs">Preprocessing and clustering 3k PBMCs</a><a id="Preprocessing-and-clustering-3k-PBMCs-1"></a><a class="docs-heading-anchor-permalink" href="#Preprocessing-and-clustering-3k-PBMCs" title="Permalink"></a></h1><p>This showcase reproduces Seurat&#39;s <a href="http://satijalab.org/seurat/pbmc3k_tutorial.html">Guided Clustering Tutorial</a></p><h2 id="Overview"><a class="docs-heading-anchor" href="#Overview">Overview</a><a id="Overview-1"></a><a class="docs-heading-anchor-permalink" href="#Overview" title="Permalink"></a></h2><ul><li><a href="#Preprocessing-and-clustering-3k-PBMCs">Preprocessing and clustering 3k PBMCs</a></li><ul><li><a href="#Overview">Overview</a></li><li><a href="#Loading-data">Loading data</a></li><li><a href="#Preprocessing">Preprocessing</a></li><li><a href="#Dimensionality-reduction">Dimensionality reduction</a></li><li><a href="#Clustering-the-cells">Clustering the cells</a></li><li><a href="#Finding-differentially-expressed-features">Finding differentially expressed features</a></li></ul></ul><h2 id="Loading-data"><a class="docs-heading-anchor" href="#Loading-data">Loading data</a><a id="Loading-data-1"></a><a class="docs-heading-anchor-permalink" href="#Loading-data" title="Permalink"></a></h2><p>The data consists of 3k PBMCs from a Healthy Donor and is freely available from <a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/1.1.0/pbmc3k">10x Genomics</a>.</p><p>You can either download the data manually</p><pre><code class="language-bash hljs">mkdir pbmc3k
wget http://cf.10xgenomics.com/samples/cell-exp/1.1.0/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz -O pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz
cd pbmc3k; tar -xzf pbmc3k_filtered_gene_bc_matrices.tar.gz --strip-components 2</code></pre><p>The <code>read_10X</code> function reads the data from the cellranger 10X pipeline and returns a labelled count matrix. Each entry indicated the number of molecules detected for each feature/gene (columns) and each cell (rows).</p><pre><code class="language-julia hljs">using Severo
X = read_10X(&quot;pbmc3k/&quot;)</code></pre><div class="admonition is-warning"><header class="admonition-header">Warning</header><div class="admonition-body"><p>The meaning of the rows and columns is different from the representation used by other packages like <a href="https://satijalab.org/seurat/">Seurat</a></p></div></div><p>Alternatively, the <code>dataset</code> function can be used to load a dataset from a predefined collection. For example, the PBMC collection is known by <code>Cell.jl</code> and can be easily loaded as follows:</p><pre><code class="language-julia hljs">using Severo
X = dataset(&quot;PBMC&quot;, &quot;3k&quot;)

# The matrix can be indexed using names or indices. For instance,
# we can look at specific genes in the first thirty cells
#X[1:30, [&quot;CD3D&quot;, &quot;TCL1A&quot;, &quot;MS4A1&quot;]]</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">2700×32738 Named sparse matrix with 2286884 Int32 nonzero entries:
	[AGATTCCTGACGAG-1,          &quot;AL627309.1&quot;]  =  1
	[CCAAAGTGCTACGA-1,          &quot;AL627309.1&quot;]  =  1
	[CCGTACACGTTGGT-1,          &quot;AL627309.1&quot;]  =  1
	[CGACTGCTTCCTCG-1,          &quot;AL627309.1&quot;]  =  1
	[CTATGTTGTCTCGC-1,          &quot;AL627309.1&quot;]  =  1
	[CTGTGAGACGAACT-1,          &quot;AL627309.1&quot;]  =  1
	[GATACTCTATCGGT-1,          &quot;AL627309.1&quot;]  =  1
	[GCTAGATGAGCTCA-1,          &quot;AL627309.1&quot;]  =  1
	[GCTTAACTACTGGT-1,          &quot;AL627309.1&quot;]  =  1
	                                           ⋮
	[TCTCAAACCTAAGC-1,            &quot;SRSF10.1&quot;]  =  1
	[TGAGACACTCAAGC-1,            &quot;SRSF10.1&quot;]  =  1
	[TGATCGGATATGCG-1,            &quot;SRSF10.1&quot;]  =  1
	[TGTAGGTGCTATGG-1,            &quot;SRSF10.1&quot;]  =  1
	[TTATCCGAGAAAGT-1,            &quot;SRSF10.1&quot;]  =  1
	[TTCAGTACCGACTA-1,            &quot;SRSF10.1&quot;]  =  1
	[TTGAATGATCTCAT-1,            &quot;SRSF10.1&quot;]  =  1
	[TTGAGGACTACGCA-1,            &quot;SRSF10.1&quot;]  =  1
	[TTTATCCTGTTGTG-1,            &quot;SRSF10.1&quot;]  =  1
	[TTTCCAGAGGTGAG-1,            &quot;SRSF10.1&quot;]  =  1</code></pre><p>The count data is stored in a sparse matrix format, only storing non-zero elements of the matrix. Any values not shown are zero.</p><article class="docstring"><header><a class="docstring-binding" id="Severo.read_data" href="#Severo.read_data"><code>Severo.read_data</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">read_data(path::AbstractString; kw...)</code></pre><p>Tries to identify and read a count matrix in any of the supported formats</p><p><strong>Arguments</strong>:</p><ul><li><code>fname</code>: path</li><li><code>kw</code>: additional keyword arguments are passed on</li></ul><p><strong>Returns values</strong>:</p><p>Returns labeled sparse matrix containing the counts</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/da4ed4b419e46bdef71284c7c291fb90c6e0e13d/src/input.jl#L711-L724">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Severo.read_10X" href="#Severo.read_10X"><code>Severo.read_10X</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">read_10X(dirname::AbstractString; unique_features=true)</code></pre><p>Read count matrix from 10X genomics</p><p><strong>Arguments</strong>:</p><ul><li><code>dirname</code>: path to directory containing matrix.mtx, genes.tsv (or features.tsv), and barcodes.tsv from 10X</li><li><code>unique_features</code>: should feature names be made unique (default: true)</li></ul><p><strong>Returns values</strong>:</p><p>Returns labeled sparse matrix containing the counts</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/da4ed4b419e46bdef71284c7c291fb90c6e0e13d/src/input.jl#L611-L624">source</a></section></article><div class="admonition is-warning"><header class="admonition-header">Missing docstring.</header><div class="admonition-body"><p>Missing docstring for <code>dataset</code>. Check Documenter&#39;s build log for details.</p></div></div><h2 id="Preprocessing"><a class="docs-heading-anchor" href="#Preprocessing">Preprocessing</a><a id="Preprocessing-1"></a><a class="docs-heading-anchor-permalink" href="#Preprocessing" title="Permalink"></a></h2><pre><code class="language- hljs">using Plots
using StatsPlots
pyplot() # hide
plot_highest_expressed(X)
savefig(&quot;he-plot.svg&quot;); nothing # hide</code></pre><p><img src="he-plot.svg" alt/></p><h3 id="Filtering"><a class="docs-heading-anchor" href="#Filtering">Filtering</a><a id="Filtering-1"></a><a class="docs-heading-anchor-permalink" href="#Filtering" title="Permalink"></a></h3><p>Two filtering function are available for filtering cells and features/genes based on basic criteria such as the number of features detected, the number of cells for which a feature is detected and the total number of counts.</p><pre><code class="language-julia hljs">X = filter_cells(X, min_features=200)
X = filter_features(X, min_cells=3)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">2700×13714 Named sparse matrix with 2282976 Int32 nonzero entries:
	[AGATTCCTGACGAG-1,          &quot;AL627309.1&quot;]  =  1
	[CCAAAGTGCTACGA-1,          &quot;AL627309.1&quot;]  =  1
	[CCGTACACGTTGGT-1,          &quot;AL627309.1&quot;]  =  1
	[CGACTGCTTCCTCG-1,          &quot;AL627309.1&quot;]  =  1
	[CTATGTTGTCTCGC-1,          &quot;AL627309.1&quot;]  =  1
	[CTGTGAGACGAACT-1,          &quot;AL627309.1&quot;]  =  1
	[GATACTCTATCGGT-1,          &quot;AL627309.1&quot;]  =  1
	[GCTAGATGAGCTCA-1,          &quot;AL627309.1&quot;]  =  1
	[GCTTAACTACTGGT-1,          &quot;AL627309.1&quot;]  =  1
	                                           ⋮
	[TCTCAAACCTAAGC-1,            &quot;SRSF10.1&quot;]  =  1
	[TGAGACACTCAAGC-1,            &quot;SRSF10.1&quot;]  =  1
	[TGATCGGATATGCG-1,            &quot;SRSF10.1&quot;]  =  1
	[TGTAGGTGCTATGG-1,            &quot;SRSF10.1&quot;]  =  1
	[TTATCCGAGAAAGT-1,            &quot;SRSF10.1&quot;]  =  1
	[TTCAGTACCGACTA-1,            &quot;SRSF10.1&quot;]  =  1
	[TTGAATGATCTCAT-1,            &quot;SRSF10.1&quot;]  =  1
	[TTGAGGACTACGCA-1,            &quot;SRSF10.1&quot;]  =  1
	[TTTATCCTGTTGTG-1,            &quot;SRSF10.1&quot;]  =  1
	[TTTCCAGAGGTGAG-1,            &quot;SRSF10.1&quot;]  =  1</code></pre><p>This filters out cells with less than 200 features/genes and features detected in less than 3 cells. For convenience, <code>filter_counts</code> combines these two into a single function, similar to Seurat. Beware that the order in which you call these functions can matter.</p><article class="docstring"><header><a class="docstring-binding" id="Severo.filter_features" href="#Severo.filter_features"><code>Severo.filter_features</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">filter_features(; min_cells = 0)</code></pre><p>Partial version of filter<em>features(A::NamedCountMatrix; min</em>cells = 0)</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/da4ed4b419e46bdef71284c7c291fb90c6e0e13d/src/macros.jl#L81-L85">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Severo.filter_cells" href="#Severo.filter_cells"><code>Severo.filter_cells</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">filter_cells(; min_features = 0, min_feature_count = 0, min_umi = 0)</code></pre><p>Partial version of filter<em>cells(A::NamedCountMatrix; min</em>features = 0, min<em>feature</em>count = 0, min_umi = 0)</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/da4ed4b419e46bdef71284c7c291fb90c6e0e13d/src/macros.jl#L81-L85">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Severo.filter_counts" href="#Severo.filter_counts"><code>Severo.filter_counts</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">filter_counts(; min_cells = 0, min_features = 0, min_feature_count = 0, min_umi = 0)</code></pre><p>Partial version of filter<em>counts(A::NamedCountMatrix; min</em>cells = 0, min<em>features = 0, min</em>feature<em>count = 0, min</em>umi = 0)</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/da4ed4b419e46bdef71284c7c291fb90c6e0e13d/src/macros.jl#L81-L85">source</a></section></article><h3 id="Normalization"><a class="docs-heading-anchor" href="#Normalization">Normalization</a><a id="Normalization-1"></a><a class="docs-heading-anchor-permalink" href="#Normalization" title="Permalink"></a></h3><p>After removing unwanted cells from the dataset, the next step is to normalize the data. The basic method normalizes the feature expression measurements for each cell by the total expression, multiplies this by a scale factor and optionally log-transforming the result.</p><pre><code class="language-julia hljs">Y = normalize_cells(X, method=:lognormalize, scale_factor=1e4)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">2700×13714 Named sparse matrix with 2282976 Float64 nonzero entries:
	[AGATTCCTGACGAG-1,          &quot;AL627309.1&quot;]  =  1.538222672557476
	[CCAAAGTGCTACGA-1,          &quot;AL627309.1&quot;]  =  1.3631966075373008
	[CCGTACACGTTGGT-1,          &quot;AL627309.1&quot;]  =  1.3945254817849748
	[CGACTGCTTCCTCG-1,          &quot;AL627309.1&quot;]  =  1.787190785781838
	[CTATGTTGTCTCGC-1,          &quot;AL627309.1&quot;]  =  1.3514100666239552
	[CTGTGAGACGAACT-1,          &quot;AL627309.1&quot;]  =  1.6848688061171004
	[GATACTCTATCGGT-1,          &quot;AL627309.1&quot;]  =  1.6623488639622728
	[GCTAGATGAGCTCA-1,          &quot;AL627309.1&quot;]  =  1.6749250073944073
	[GCTTAACTACTGGT-1,          &quot;AL627309.1&quot;]  =  1.8821985210730219
	                                           ⋮
	[TCTCAAACCTAAGC-1,            &quot;SRSF10.1&quot;]  =  2.050332029149289
	[TGAGACACTCAAGC-1,            &quot;SRSF10.1&quot;]  =  1.873284442922057
	[TGATCGGATATGCG-1,            &quot;SRSF10.1&quot;]  =  2.021955420152209
	[TGTAGGTGCTATGG-1,            &quot;SRSF10.1&quot;]  =  1.2128182558075358
	[TTATCCGAGAAAGT-1,            &quot;SRSF10.1&quot;]  =  1.2133165770732028
	[TTCAGTACCGACTA-1,            &quot;SRSF10.1&quot;]  =  1.529100016033425
	[TTGAATGATCTCAT-1,            &quot;SRSF10.1&quot;]  =  1.7863632500815836
	[TTGAGGACTACGCA-1,            &quot;SRSF10.1&quot;]  =  0.9465443048463601
	[TTTATCCTGTTGTG-1,            &quot;SRSF10.1&quot;]  =  1.31322083398467
	[TTTCCAGAGGTGAG-1,            &quot;SRSF10.1&quot;]  =  1.7178390681467346</code></pre><article class="docstring"><header><a class="docstring-binding" id="Severo.normalize_cells" href="#Severo.normalize_cells"><code>Severo.normalize_cells</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">normalize_cells(; method = :lognormalize, scale_factor::Real = 1.0, dtype::Type{T} = Float64) where T &lt;: AbstractFloat</code></pre><p>Partial version of normalize<em>cells(X::NamedCountMatrix; method = :lognormalize, scale</em>factor::Real = 1.0, dtype::Type{T} = Float64) where T &lt;: AbstractFloat</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/da4ed4b419e46bdef71284c7c291fb90c6e0e13d/src/macros.jl#L81-L85">source</a></section></article><h3 id="Identifying-highly-variable-features"><a class="docs-heading-anchor" href="#Identifying-highly-variable-features">Identifying highly variable features</a><a id="Identifying-highly-variable-features-1"></a><a class="docs-heading-anchor-permalink" href="#Identifying-highly-variable-features" title="Permalink"></a></h3><p>Next, we select a subset of features that exhibit high cell-to-cell variation in the dataset (i.e, they are highly expressed in some cells, and lowly expressed in others).</p><pre><code class="language-julia hljs">hvf = find_variable_features(X, 2000; method=:vst)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">2000-element Named SubArray{Int64, 1, Vector{Int64}, Tuple{UnitRange{Int64}}, true}
features              │ 
──────────────────────┼──────
&quot;PPBP&quot;                │  3316
&quot;LYZ&quot;                 │  8682
&quot;S100A9&quot;              │   908
&quot;IGLL5&quot;               │ 13275
&quot;GNLY&quot;                │  1804
&quot;FTL&quot;                 │ 12944
&quot;PF4&quot;                 │  3315
&quot;FTH1&quot;                │  7755
⋮                           ⋮
&quot;EIF3D&quot;               │ 13383
&quot;POLR2I&quot;              │ 12681
&quot;ZCCHC10&quot;             │  3924
&quot;SEC61B&quot;              │  6669
&quot;RP11-164H13.1&quot;       │  9585
&quot;NELL2&quot;               │  8455
&quot;GTF2H2&quot;              │  3747
&quot;NF1&quot;                 │ 11047</code></pre><article class="docstring"><header><a class="docstring-binding" id="Severo.find_variable_features" href="#Severo.find_variable_features"><code>Severo.find_variable_features</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">find_variable_features(nfeatures = 2000; method = :vst, dtype::Type{&lt;:AbstractFloat} = Float64, kw...)</code></pre><p>Partial version of find<em>variable</em>features(counts::NamedCountMatrix, nfeatures = 2000; method = :vst, dtype::Type{&lt;:AbstractFloat} = Float64, kw...)</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/da4ed4b419e46bdef71284c7c291fb90c6e0e13d/src/macros.jl#L81-L85">source</a></section></article><h2 id="Dimensionality-reduction"><a class="docs-heading-anchor" href="#Dimensionality-reduction">Dimensionality reduction</a><a id="Dimensionality-reduction-1"></a><a class="docs-heading-anchor-permalink" href="#Dimensionality-reduction" title="Permalink"></a></h2><h3 id="Scaling"><a class="docs-heading-anchor" href="#Scaling">Scaling</a><a id="Scaling-1"></a><a class="docs-heading-anchor-permalink" href="#Scaling" title="Permalink"></a></h3><p>Prior to dimensional reduction techniques like PCA, it&#39;s a good idea to <code>scale_features</code> the data. Scaling performs two basic operations:</p><ul><li>Shifts the expression of each gene, so that the mean expression across cells is 0</li><li>Scales the expression of each gene, so that the standard deviation across cells is 1. This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate</li><li>Clips values exceeding standard deviation of <code>scale_max</code></li></ul><pre><code class="language-julia hljs">Y = Y[:,hvf] # only use highly-variable features
S = scale_features(Y; scale_max=10)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">CenteredMatrix:
  A = 2700×2000 Named sparse matrix with 515841 Float64 nonzero entries:
	[AAACCGTGCTTCCG-1,                &quot;PPBP&quot;]  =  3.05759
	[AAATCAACCCTATT-1,                &quot;PPBP&quot;]  =  1.983
	[AAATCCCTGCTATG-1,                &quot;PPBP&quot;]  =  5.03146
	[AACCGCCTCTACGA-1,                &quot;PPBP&quot;]  =  2.44748
	[AAGCCTGACATGCA-1,                &quot;PPBP&quot;]  =  3.68051
	[AATCCTTGACGGGA-1,                &quot;PPBP&quot;]  =  3.12117
	[ACCCACTGGTTCAG-1,                &quot;PPBP&quot;]  =  10.143
	[ACCTGAGATATCGG-1,                &quot;PPBP&quot;]  =  9.63048
	[ACCTGGCTGTCTTT-1,                &quot;PPBP&quot;]  =  7.59301
	                                           ⋮
	[TCAAGGACAGCGTT-1,                 &quot;NF1&quot;]  =  8.53908
	[TCGAGAACGTTAGC-1,                 &quot;NF1&quot;]  =  6.14793
	[TGACGCCTGTACCA-1,                 &quot;NF1&quot;]  =  4.82082
	[TGATAAACTCCGTC-1,                 &quot;NF1&quot;]  =  7.88764
	[TGCAATCTTCAGGT-1,                 &quot;NF1&quot;]  =  4.24277
	[TGGAGACTATCAGC-1,                 &quot;NF1&quot;]  =  6.24035
	[TGTAGGTGCTATGG-1,                 &quot;NF1&quot;]  =  5.4219
	[TTAGTCTGTGCACA-1,                 &quot;NF1&quot;]  =  8.64597
	[TTGCTATGGTACGT-1,                 &quot;NF1&quot;]  =  6.43414
	[TTGGAGACGCTATG-1,                 &quot;NF1&quot;]  =  6.97187
  mu = 2000-element Named Vector{Float64}
features              │ 
──────────────────────┼──────────
&quot;PPBP&quot;                │  0.142967
&quot;LYZ&quot;                 │  0.972527
&quot;S100A9&quot;              │  0.645189
&quot;IGLL5&quot;               │   0.18421
&quot;GNLY&quot;                │  0.407293
&quot;FTL&quot;                 │   2.94261
&quot;PF4&quot;                 │  0.110369
&quot;FTH1&quot;                │   3.22873
⋮                               ⋮
&quot;EIF3D&quot;               │  0.775623
&quot;POLR2I&quot;              │  0.396716
&quot;ZCCHC10&quot;             │  0.211964
&quot;SEC61B&quot;              │  0.656028
&quot;RP11-164H13.1&quot;       │ 0.0889054
&quot;NELL2&quot;               │   0.25572
&quot;GTF2H2&quot;              │  0.131466
&quot;NF1&quot;                 │  0.130034</code></pre><article class="docstring"><header><a class="docstring-binding" id="Severo.scale_features" href="#Severo.scale_features"><code>Severo.scale_features</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">scale_features(; scale_max::Real = Inf, dtype::Type{&lt;:AbstractFloat} = Float64)</code></pre><p>Partial version of scale<em>features(X::(NamedArray{T, 2, SparseMatrixCSC{T, Int64}} where T); scale</em>max::Real = Inf, dtype::Type{&lt;:AbstractFloat} = Float64)</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/da4ed4b419e46bdef71284c7c291fb90c6e0e13d/src/macros.jl#L81-L85">source</a></section></article><h3 id="Principal-component-analysis"><a class="docs-heading-anchor" href="#Principal-component-analysis">Principal component analysis</a><a id="Principal-component-analysis-1"></a><a class="docs-heading-anchor-permalink" href="#Principal-component-analysis" title="Permalink"></a></h3><p>Reduce the dimensionality of the data by running principal component analysis (PCA), which reveals the main axes of variation and de-noises the data.</p><pre><code class="language-julia hljs">em = embedding(S, 15, method=:pca)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">PC-1
Positive: WeakRefStrings.String31[&quot;CST3&quot;, &quot;TYROBP&quot;, &quot;LST1&quot;, &quot;AIF1&quot;, &quot;FTL&quot;, &quot;FCN1&quot;, &quot;LYZ&quot;, &quot;FTH1&quot;, &quot;FCER1G&quot;, &quot;S100A9&quot;]
Negative: WeakRefStrings.String31[&quot;LTB&quot;, &quot;IL32&quot;, &quot;CD2&quot;, &quot;ACAP1&quot;, &quot;STK17A&quot;, &quot;CTSW&quot;, &quot;CD247&quot;, &quot;GIMAP5&quot;, &quot;AQP3&quot;, &quot;CCL5&quot;]
PC-2
Positive: WeakRefStrings.String31[&quot;CD79A&quot;, &quot;MS4A1&quot;, &quot;TCL1A&quot;, &quot;HLA-DRA&quot;, &quot;HLA-DQA1&quot;, &quot;HLA-DQB1&quot;, &quot;LINC00926&quot;, &quot;CD79B&quot;, &quot;HLA-DRB1&quot;, &quot;CD74&quot;]
Negative: WeakRefStrings.String31[&quot;NKG7&quot;, &quot;PRF1&quot;, &quot;CST7&quot;, &quot;GZMA&quot;, &quot;GZMB&quot;, &quot;FGFBP2&quot;, &quot;CTSW&quot;, &quot;GNLY&quot;, &quot;GZMH&quot;, &quot;SPON2&quot;]
PC-3
Positive: WeakRefStrings.String31[&quot;HLA-DQA1&quot;, &quot;CD79A&quot;, &quot;HLA-DQB1&quot;, &quot;CD79B&quot;, &quot;HLA-DPB1&quot;, &quot;CD74&quot;, &quot;HLA-DPA1&quot;, &quot;MS4A1&quot;, &quot;HLA-DRB1&quot;, &quot;HLA-DRB5&quot;]
Negative: WeakRefStrings.String31[&quot;PPBP&quot;, &quot;PF4&quot;, &quot;SDPR&quot;, &quot;SPARC&quot;, &quot;GNG11&quot;, &quot;NRGN&quot;, &quot;GP9&quot;, &quot;RGS18&quot;, &quot;TUBB1&quot;, &quot;HIST1H2AC&quot;]
PC-4
Positive: WeakRefStrings.String31[&quot;HLA-DQA1&quot;, &quot;CD79A&quot;, &quot;CD79B&quot;, &quot;HLA-DQB1&quot;, &quot;MS4A1&quot;, &quot;CD74&quot;, &quot;HLA-DPB1&quot;, &quot;HIST1H2AC&quot;, &quot;PF4&quot;, &quot;HLA-DQA2&quot;]
Negative: WeakRefStrings.String31[&quot;RPS2&quot;, &quot;VIM&quot;, &quot;S100A6&quot;, &quot;S100A8&quot;, &quot;S100A4&quot;, &quot;IL32&quot;, &quot;S100A9&quot;, &quot;TMSB10&quot;, &quot;GIMAP7&quot;, &quot;S100A10&quot;]
PC-5
Positive: WeakRefStrings.String31[&quot;GZMB&quot;, &quot;FGFBP2&quot;, &quot;NKG7&quot;, &quot;GNLY&quot;, &quot;PRF1&quot;, &quot;CCL4&quot;, &quot;CST7&quot;, &quot;SPON2&quot;, &quot;GZMA&quot;, &quot;GZMH&quot;]
Negative: WeakRefStrings.String31[&quot;LTB&quot;, &quot;VIM&quot;, &quot;AQP3&quot;, &quot;PPA1&quot;, &quot;MAL&quot;, &quot;RPS2&quot;, &quot;KIAA0101&quot;, &quot;CD2&quot;, &quot;CORO1B&quot;, &quot;CYTIP&quot;]
</code></pre><pre><code class="language- hljs">plot_elbow(em)
savefig(&quot;elbow-plot.svg&quot;); nothing # hide</code></pre><p><img src="elbow-plot.svg" alt/></p><article class="docstring"><header><a class="docstring-binding" id="Severo.embedding" href="#Severo.embedding"><code>Severo.embedding</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">embedding(ncomponents::Int64 = 50; method = :pca, kw...)</code></pre><p>Partial version of embedding(X, ncomponents::Int64 = 50; method = :pca, kw...)</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/da4ed4b419e46bdef71284c7c291fb90c6e0e13d/src/macros.jl#L81-L85">source</a></section></article><h2 id="Clustering-the-cells"><a class="docs-heading-anchor" href="#Clustering-the-cells">Clustering the cells</a><a id="Clustering-the-cells-1"></a><a class="docs-heading-anchor-permalink" href="#Clustering-the-cells" title="Permalink"></a></h2><p>Briefly, the clustering embed cells in a graph structure with edges between cells with similar feature expression patterns, and then attempt to partition this graph into highly interconnected &#39;quasi-cliques&#39; or &#39;communities&#39;.</p><h3 id="Computing-the-neighborhood-graph"><a class="docs-heading-anchor" href="#Computing-the-neighborhood-graph">Computing the neighborhood graph</a><a id="Computing-the-neighborhood-graph-1"></a><a class="docs-heading-anchor-permalink" href="#Computing-the-neighborhood-graph" title="Permalink"></a></h3><p>First, we need to find the k-nearest neighbors of all the cells based on the euclidean distance in PCA space, and then compute a graph based on the shared overlap in the local neighborhoods between any two cells (Jaccard similarity). These step are performed using the <code>nearest_neighbours</code> and <code>jaccard_index</code> functions. For convenience, the two steps are also combined in the <code>shared_nearest_neighours</code> function.</p><pre><code class="language-julia hljs">nn = nearest_neighbours(em, 20, dims=1:10)
snn = jaccard_index(nn, prune=1/15)
# or as a single step
#snn = nearest_neighbours(em, 20, dims=1:10, prune=1/15)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">2700×2700 Named sparse matrix with 198324 Float64 nonzero entries:
	[AAACATACAACCAC-1, AAACATACAACCAC-1]  =  1.0
	[AAATTGACTCGCTC-1, AAACATACAACCAC-1]  =  0.1111111111111111
	[AACTCTTGCAGGAG-1, AAACATACAACCAC-1]  =  0.08108108108108109
	[AATCTAGAATCGGT-1, AAACATACAACCAC-1]  =  0.08108108108108109
	[ACCACGCTGCTGTA-1, AAACATACAACCAC-1]  =  0.14285714285714285
	[ACGCCGGAAATGCC-1, AAACATACAACCAC-1]  =  0.17647058823529413
	[ACTAGGTGGAACTC-1, AAACATACAACCAC-1]  =  0.08108108108108109
	[AGCCGTCTGAGAGC-1, AAACATACAACCAC-1]  =  0.08108108108108109
	[AGCGGGCTTGCCAA-1, AAACATACAACCAC-1]  =  0.1111111111111111
	                                      ⋮
	[TCGGTAGAGTAGGG-1, TTTGCATGCCTCAC-1]  =  0.08108108108108109
	[TGATAAACTTTCAC-1, TTTGCATGCCTCAC-1]  =  0.08108108108108109
	[TGATACCTGTTGGT-1, TTTGCATGCCTCAC-1]  =  0.08108108108108109
	[TGCCAGCTTGGCAT-1, TTTGCATGCCTCAC-1]  =  0.14285714285714285
	[TGGTTACTGTTCTT-1, TTTGCATGCCTCAC-1]  =  0.08108108108108109
	[TGTAACCTAGAGGC-1, TTTGCATGCCTCAC-1]  =  0.1111111111111111
	[TTAGAATGTGGTGT-1, TTTGCATGCCTCAC-1]  =  0.08108108108108109
	[TTAGCTACAACCGT-1, TTTGCATGCCTCAC-1]  =  0.1111111111111111
	[TTCGATTGAGCATC-1, TTTGCATGCCTCAC-1]  =  0.14285714285714285
	[TTTGCATGCCTCAC-1, TTTGCATGCCTCAC-1]  =  1.0</code></pre><article class="docstring"><header><a class="docstring-binding" id="Severo.nearest_neighbours" href="#Severo.nearest_neighbours"><code>Severo.nearest_neighbours</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">nearest_neighbours(k::Int64; dims = (:), metric::SemiMetric = Euclidean(), include_self::Bool = true, ntables::Int64 = 2 * size(X, 2)) where T</code></pre><p>Partial version of nearest<em>neighbours(X::NamedArray{T, 2}, k::Int64; dims = (:), metric::SemiMetric = Euclidean(), include</em>self::Bool = true, ntables::Int64 = 2 * size(X, 2)) where T</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/da4ed4b419e46bdef71284c7c291fb90c6e0e13d/src/macros.jl#L81-L85">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Severo.jaccard_index" href="#Severo.jaccard_index"><code>Severo.jaccard_index</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">jaccard_index(k::Int64; prune::Real = 1 / 15, dtype::Type{R} = Float64) where R &lt;: AbstractFloat</code></pre><p>Partial version of jaccard_index(nn::(NamedArray{T, 2} where T), k::Int64; prune::Real = 1 / 15, dtype::Type{R} = Float64) where R &lt;: AbstractFloat</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/da4ed4b419e46bdef71284c7c291fb90c6e0e13d/src/macros.jl#L81-L85">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Severo.shared_nearest_neighbours" href="#Severo.shared_nearest_neighbours"><code>Severo.shared_nearest_neighbours</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">shared_nearest_neighbours(k::Int64; kw...) where T</code></pre><p>Partial version of shared<em>nearest</em>neighbours(X::NamedArray{T, 2}, k::Int64; kw...) where T</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/da4ed4b419e46bdef71284c7c291fb90c6e0e13d/src/macros.jl#L81-L85">source</a></section></article><h3 id="Clustering-the-neighborhood-graph"><a class="docs-heading-anchor" href="#Clustering-the-neighborhood-graph">Clustering the neighborhood graph</a><a id="Clustering-the-neighborhood-graph-1"></a><a class="docs-heading-anchor-permalink" href="#Clustering-the-neighborhood-graph" title="Permalink"></a></h3><p>Clustering is performed using a graph-based method called <code>Louvain clustering</code> which tries to detect communities based on optimizing a modularity function. Since this is a greedy algorithm, which can get stuck in local minima based on random initialization, multiple restarts are required to find a more &quot;global&quot; minima.</p><pre><code class="language-julia hljs">lbls = cluster(snn)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">2700-element Named Vector{Int64}
cells            │ 
─────────────────┼───
AAACATACAACCAC-1 │  1
AAACATTGAGCTAC-1 │  2
AAACATTGATCAGC-1 │  1
AAACCGTGCTTCCG-1 │  3
AAACCGTGTATGCG-1 │  4
AAACGCACTGGTAC-1 │  1
AAACGCTGACCAGT-1 │  5
AAACGCTGGTTCTT-1 │  5
⋮                   ⋮
TTTCAGTGTGCAGT-1 │  2
TTTCCAGAGGTGAG-1 │  1
TTTCGAACACCTGA-1 │ 10
TTTCGAACTCTCAT-1 │  8
TTTCTACTGAGGCA-1 │  2
TTTCTACTTCCTCG-1 │  2
TTTGCATGAGAGGC-1 │  2
TTTGCATGCCTCAC-1 │  7</code></pre><article class="docstring"><header><a class="docstring-binding" id="Severo.cluster" href="#Severo.cluster"><code>Severo.cluster</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">cluster(; kw...)</code></pre><p>Partial version of cluster(SNN::NeighbourGraph; kw...)</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/da4ed4b419e46bdef71284c7c291fb90c6e0e13d/src/macros.jl#L81-L85">source</a></section></article><h2 id="Finding-differentially-expressed-features"><a class="docs-heading-anchor" href="#Finding-differentially-expressed-features">Finding differentially expressed features</a><a id="Finding-differentially-expressed-features-1"></a><a class="docs-heading-anchor-permalink" href="#Finding-differentially-expressed-features" title="Permalink"></a></h2><p>We can find markers that define clusters via differential expression. It identifies positive and negative markers for a single cluster compared to the other cells. <code>find_markers</code> performs this process for every cluster. The function returns a <code>DataFrame</code> containing for each marker: the p-value, the log-foldchange and statistical score.</p><pre><code class="language-julia hljs">dx = find_markers(X, lbls; method=:wilcoxon)
dx = filter_rank_markers(dx)
dx[1:10, :]</code></pre><div class="data-frame"><p>10 rows × 5 columns</p><table class="data-frame"><thead><tr><th></th><th>group</th><th>score</th><th>pval</th><th>logfc</th><th>feature</th></tr><tr><th></th><th title="Int64">Int64</th><th title="Float64">Float64</th><th title="Float64">Float64</th><th title="Float64">Float64</th><th title="WeakRefStrings.String31">String3…</th></tr></thead><tbody><tr><th>1</th><td>1</td><td>15.6115</td><td>6.07706e-55</td><td>-1.67376</td><td>HLA-DRB1</td></tr><tr><th>2</th><td>1</td><td>15.225</td><td>2.41296e-52</td><td>-2.21502</td><td>TYROBP</td></tr><tr><th>3</th><td>1</td><td>15.2236</td><td>2.46701e-52</td><td>-2.50869</td><td>HLA-DRA</td></tr><tr><th>4</th><td>1</td><td>14.6119</td><td>2.35899e-48</td><td>-1.66346</td><td>FCER1G</td></tr><tr><th>5</th><td>1</td><td>14.5481</td><td>6.00609e-48</td><td>-1.07385</td><td>HLA-DRB5</td></tr><tr><th>6</th><td>1</td><td>14.2512</td><td>4.40701e-46</td><td>-1.64251</td><td>HLA-DPA1</td></tr><tr><th>7</th><td>1</td><td>13.7023</td><td>9.83279e-43</td><td>-1.64646</td><td>HLA-DPB1</td></tr><tr><th>8</th><td>1</td><td>12.6029</td><td>2.03398e-36</td><td>-1.95414</td><td>CD74</td></tr><tr><th>9</th><td>1</td><td>12.3178</td><td>7.26175e-35</td><td>-0.693704</td><td>HLA-DMA</td></tr><tr><th>10</th><td>1</td><td>12.1182</td><td>8.45788e-34</td><td>-0.847469</td><td>CFD</td></tr></tbody></table></div><p>Performance can be improved by first filtering the features based on the percentage of cells and the fold-change between the groups.</p><pre><code class="language-julia hljs"># filtering and DE on the lognormalized data
sel = prefilter_markers(Y, lbls, logfc_threshold=0.25, min_pct=0.1, log=true, only_pos=true)
dx = find_markers(Y, lbls; method=:t, log=true, selection=sel) # t-test
dx = filter_rank_markers(dx)
dx[1:10, :]</code></pre><div class="data-frame"><p>10 rows × 5 columns</p><table class="data-frame"><thead><tr><th></th><th>group</th><th>score</th><th>pval</th><th>logfc</th><th>feature</th></tr><tr><th></th><th title="Int64">Int64</th><th title="Float64">Float64</th><th title="Float64">Float64</th><th title="Float64">Float64</th><th title="String">String</th></tr></thead><tbody><tr><th>1</th><td>1</td><td>30.4872</td><td>3.03125e-151</td><td>1.20855</td><td>IL32</td></tr><tr><th>2</th><td>1</td><td>28.8907</td><td>5.12713e-141</td><td>1.26314</td><td>LTB</td></tr><tr><th>3</th><td>1</td><td>15.6553</td><td>1.57264e-47</td><td>1.22055</td><td>CD2</td></tr><tr><th>4</th><td>1</td><td>14.3902</td><td>3.71724e-43</td><td>0.487126</td><td>VIM</td></tr><tr><th>5</th><td>1</td><td>14.2749</td><td>1.64055e-41</td><td>0.604693</td><td>GIMAP7</td></tr><tr><th>6</th><td>1</td><td>12.6808</td><td>7.23751e-33</td><td>1.23004</td><td>AQP3</td></tr><tr><th>7</th><td>1</td><td>11.6523</td><td>3.66613e-29</td><td>0.601767</td><td>ANXA1</td></tr><tr><th>8</th><td>1</td><td>10.472</td><td>8.95668e-24</td><td>0.842795</td><td>TRADD</td></tr><tr><th>9</th><td>1</td><td>9.15615</td><td>6.08564e-19</td><td>0.768018</td><td>TNFAIP8</td></tr><tr><th>10</th><td>1</td><td>8.69457</td><td>3.0797e-17</td><td>0.718181</td><td>MAL</td></tr></tbody></table></div><article class="docstring"><header><a class="docstring-binding" id="Severo.find_markers" href="#Severo.find_markers"><code>Severo.find_markers</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">find_markers(X::Union{NamedCountMatrix, NamedDataMatrix}, idents::NamedVector{&lt;:Integer};
    method=:wilcoxon, selection::Union{Nothing, NamedArray{Bool, 2}, AbstractArray{Bool,2}}=nothing, log::Bool=false, kw...)</code></pre><p>Finds markers (differentially expressed genes) for each of the classes in a dataset.</p><p><strong>Arguments</strong>:</p><pre><code class="nohighlight hljs">-`X`: count or data matrix
-`idents`: class identity for each cell
-`method`: Which test to use, supported are: [wilcoxon, t]
-`selection`: a selection of features and groups that should be considered
-`log`: the data is in log-scale (default = false)
-`kw...`: additional parameters passed down to the method</code></pre><p><strong>Return values</strong>:</p><p>A <code>DataFrame</code> containing a list of putative markers with associated statistics (p-values and scores) and log fold-changes.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/da4ed4b419e46bdef71284c7c291fb90c6e0e13d/src/diffexpr.jl#L192-L210">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Severo.filter_rank_markers" href="#Severo.filter_rank_markers"><code>Severo.filter_rank_markers</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">filter_rank_markers(de::DataFrame; pval_thresh::Real=1e-2, ngenes::Integer=typemax(Int64))</code></pre><p>Filters and ranks a list of markers (differentially expressed genes).</p><p><strong>Arguments</strong>:</p><pre><code class="nohighlight hljs">-`de`: list of markers returned by [find_markers](@ref)
-`pval_thresh`: only keep markers with pval &lt; pval_thresh
-`count`: the number of highest-ranked markers to keep
-`rankby_abs`: rank based on absolute value of the scores</code></pre><p><strong>Return values</strong>:</p><p>A <code>DataFrame</code> containing a ranked list of filtered markers.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/da4ed4b419e46bdef71284c7c291fb90c6e0e13d/src/diffexpr.jl#L320-L335">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="Severo.prefilter_markers" href="#Severo.prefilter_markers"><code>Severo.prefilter_markers</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">prefilter_markers(X::Union{NamedCountMatrix, NamedDataMatrix}, idents::NamedVector{&lt;:Integer};
        logfc_threshold::Real=0.0, min_pct::Real=0.0, min_diff_pct::Real=-Inf, only_pos:Bool=false, log::Bool=false)</code></pre><p>Filter features for each of the classes in a dataset.</p><p><strong>Arguments</strong>:</p><pre><code class="nohighlight hljs">-`X`: count or data matrix
-`idents`: class identity for each cell
-`logfc_threshold`: Limit testing to features which show, on average, at least X-fold difference (log-scale) between the two groups of cells
-`min_pct`: only test features that are detected in a minimum fraction of `min_pct` cells in either of the two populations
-`min_diff_pct`: only test features that show a minimum difference in the fraction of detection between the two groups.
-`only_pos`: only return features with positive log fold-change
-`log`: the data is in log-scale (default = false)</code></pre><p><strong>Return values</strong>:</p><p>Selection matrix for each feature and class</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ExaScience/Severo.jl/blob/da4ed4b419e46bdef71284c7c291fb90c6e0e13d/src/diffexpr.jl#L121-L140">source</a></section></article><pre><code class="language- hljs">plot_violin(X, [&quot;CST3&quot;, &quot;NKG7&quot;, &quot;PPBP&quot;, &quot;S100A4&quot;], lbls)
savefig(&quot;violin-plot.svg&quot;); nothing # hide</code></pre><p><img src="violin-plot.svg" alt/></p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../">« Home</a><a class="docs-footer-nextpage" href="../performance/">Performance »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.6 on <span class="colophon-date" title="Sunday 26 September 2021 11:40">Sunday 26 September 2021</span>. Using Julia version 1.6.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
